---
title: "admixr - R package for facilitating ADMIXTOOLS analyses"
author:
#date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
bibliography: library.bib
csl: bioinformatics.csl
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
library(ggplot2)
library(admixr)
path_to_eigenstrat <- file.path(system.file(package = "admixr", "extdata"), "snps")
```

## Abstract

We present a new R package _admixr_, which provides a convenient interface for
the ADMIXTOOLS software suite. In a traditional ADMIXTOOLS workflow, the user
must first generate a set of text configuration files tailored to each
individual analysis, often using a combination shell scripting and manual
editing. After executing an ADMIXTOOLS program on the command line using the
provided configuration files, generated log files need to be parsed to extract
values of interest before they can be used in a downstream analysis. Our
package simplifies this process by automating all low-level configuration and
parsing steps without any user intervention, making the previously complex
workflow as simple as running a single line of R code. Furthermore, by unifying
all steps of an analytic workflow under a single framework, this package allows
building of fully automated analytic pipelines, significantly improving the
reproducibility of population genetic studies.

**Source code and documentation:** Source code, installation instructions,
documentation and a tutorial can be found on the package website at
[http://bodkan.net/admixr](http://bodkan.net/admixr).

## Introduction

The growing numbers of available ancient and modern DNA samples have
transformed our understanding of the evolutionary history of humans and other
species. Several statistical methods have been developed to make inferences
about past population movements and admixture events from genomic data
[@racimo_evidence_2015]. Chief among those has been a series of admixture
statistics such as $f_3$, $f_4$, $f_4$-ratio and $D$ as implemented in the set
of command-line utilities distributed in the ADMIXTOOLS software package
[@patterson_ancient_2012]. Although instrumental in making inferences about
ancient and modern human history - indeed, they have been used in nearly all
recent studies of ancient DNA in high-profile scientific journals
[@fu_genetic_2016; @hajdinjak_reconstructing_2018; @lazaridis_genomic_2016] -
ADMIXTOOLS utilities suffer from several disadvantages. First, each individual
analysis or hypothesis test rely on a set of text configuration files which
have to be generated using a combination of shell scripting and manual editing.
Furthermore, after running an ADMIXTOOLS command on the command-line, the user
needs to extract relevant values from a log file before they can be imported in
another software for further analysis and plotting. This workflow is quite
cumbersome and error-prone, especially if the user needs to quickly iterate
through complex hypotheses involving many different populations or samples.
Most importantly, however, any reliance on manual user intervention for both
the preparation and parsing of configuration and output log files makes it
difficult to follow the rules of good practice of reproducible research. To
overcome these challenges, we present a new R package which provides a unified
interface for population admixture analyses, utilizing ADMIXTOOLS software
suite for the underlying calculations. The package completely automates the
generation, processing and parsing of all intermediate files, hiding all
low-level details from the users and allowing them to focus on the analysis
itself. Importantly, in addition to providing a more user-friendly interface,
unifying the entire analytic workflow in an R environment makes it possible to
implement fully automated reproducible analytic pipelines.


## Implementation and example usage

The _admixr_ package is implemented using the R programming language
[@r_core_2018]. It's core functionality includes the following set of R
functions: `f3()`, `d()`, `f4()`, `f4ratio()` and `qpAdm()`, each corresponding
to one of the population admixture statistics implemented in ADMIXTOOLS.
Traditionally, performing even the most trivial analysis using the ADMIXTOOLS
package presents a significant amount of overhead for the user. For example, in
order to calculate $f_4$-ratio statistic such as

$$f_4(\textrm{Altai}, \textrm{Chimp}; \textrm{X}, \textrm{Mbuti}) /
  f_4(\textrm{Altai}, \textrm{Chimp}; \textrm{Vindija}, \textrm{Mbuti}),$$

that is, to estimate the proportion of Neandertal ancestry in a set of
individuals $X$, the user needs to create parameter files and population list
files, tailored for this specific analysis, run the `qpF4ratio` program on the
command-line, and to parse the log file to obtain relevant information (see
Supplementary Information for an example of performing the complete analysis
using a traditional ADMIXTOOLS command-line-based workflow). Note that changing
the analysis setup in a slightly different way or including a different set of
samples requires re-generating all files again.

In contrast, the same analysis using the _admixr_ package can be performed
using just the following single R command:

```{r}
result <- f4ratio(
    X = c("French", "Han", "Papuan", "Yoruba"),
    A = "Altai", B = "Vindija", C = "Mbuti", O = "Chimp",
    prefix = path_to_eigenstrat
)
```

Internally, this code generates all necessary configuration files, runs the
`qpF4ratio` command, parses the output file, and presents the user with the
following R `data.frame` object for further statistical analysis an plotting:

```{r, echo = FALSE}
knitr::kable(result)
```

Other ADMIXTOOLS analyses can be performed in a similar way, and are described
in the tutorial R vignette in a more detail.


## Additional functionality

The fact that ADMIXTOOLS requires data to be in a peculiar EIGENSTRAT file
format makes data processing and filtering can be challenging because this
format is not supported by most of the standard bioinformatic tools. To
simplify handling and processing of EIGENSTRAT datasets, _admixr_ implements a
set of complementary functions that can be useful during data-preparation:

* Filtering of SNP data based on a given BED file &mdash; `filter_sites()`.
* Ancient DNA damage or GC content filtering &mdash; `filter_damage()`, `filter_gc()`.
* Merging of EIGENSTRAT datasets &mdash; `merge_eigenstrat()`.
* Calculating coverage of sites in each sample &mdash; `count_snps()`.
* Merging of individual population labels into larger groups &mdash; `group_labels()`. 
* VCF-EIGENSTRAT format conversion &mdash; `vcf_to_eigenstrat()`, `eigenstrat_to_vcf()`.

## Funding

Development of this software was supported by the Max Planck Society.

## References
