---
title: "admixr - tutorial"
author: "Martin Petr"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```







This vignette describes how to calculate various population admixture statistics
($D$, $f_4$, etc.) using the `admixr` package. It assumes that you already
know a little bit about them, because we don't have enough space
to go into details of how and why they work. If you want to learn more and
are not afraid of little bit of math, I highly recommend Benjamin Peter's
[wonderful overview](http://www.genetics.org/content/early/2016/02/03/genetics.115.183913)
of the subject and Nick Patterson's [original ADMIXTOOLS paper](http://www.genetics.org/content/192/3/1065).



## Introduction

[ADMIXTOOLS](http://www.genetics.org/content/192/3/1065) is a widely used
software package for calculating admixture statistics and testing population
admixture hypotheses. However, although powerful and comprehensive, it is not
really user-friendly.

A typical ADMIXTOOLS workflow usually involves a combination of `sed`/`awk`/shell
scripting and manual editing to create different configuration files. These are then
passed as command-line arguments to one of ADMIXTOOLS' commands and control how
to run a particular analysis. The results are then redirected to
another file, which has to be parsed by the user to extract values of interest,
often using command-line utilities again or (worse) by manual copy-pasting.
The processed results are then analysed in R, Excel or another program.

This workflow is very cumbersome, especially if one wants to explore many hypotheses
involving different combinations of populations. Most importantly, however, it
makes it difficult to follow good practices of reproducible science, as it is
nearly impossible to construct reproducible automated "pipelines".

This R package makes it possible to perform all stages of an ADMIXTOOLS
analysis entirely from within R. It provides a set of convenient functions that
completely remove the need for "low level" configuration of individual
ADMIXTOOLS programs, allowing users to focus on the analysis itself.










## Installation

To install `admixr` from GitHub you need to install the package 
`devtools` first. To do this, you can simply run (in R):

```{r, eval = FALSE}
install.packages("devtools")
devtools::install_github("bodkan/admixr")
```

Furthermore, if you want to follow the examples in this vignette, you will need the [tidyverse](https://www.tidyverse.org) collection of
packages for convenient data analysis, which you can install with:

```{r, results = "hide", message = FALSE}
# specifying 'repos' to avoid Travis CI build errors - not needed in user code
install.packages("tidyverse", repos = "http://cran.us.r-project.org")
```

When everything is ready, you can run:

```{r}
library(admixr)
library(tidyverse)
```

**Note that in order to run `admixr` analyses, you need a working
installation of ADMIXTOOLS!** Explaining how to compile it is beyond
the scope of this document, but you can find help 
[here](https://github.com/DReichLab/AdmixTools). **Furthermore, you
need to make sure that R can find ADMIXTOOLS binaries on the `$PATH`.** 
If this is not the case, running `library(admixr)` will show a warning message
with instructions on how to fix this.







## A note about EIGENSTRAT files

If you have an EIGENSTRAT "triplet" of files ready, and just want to know
how to calculate different admixture statistics, feel free to skip
this section.

### EIGENSTRAT file format

ADMIXTOOLS software uses a peculiar set of genetic file formats, which 
seem strange if you are used to working with
[VCF files](http://samtools.github.io/hts-specs/VCFv4.3.pdf).
However, the basic idea remains the same - we want to store and
access SNP data (REF/ALT alleles) of a set of individuals at a defined 
set of genomic positions.

EIGENSTRAT data sets always contain three kinds of files:

- `ind` file - specifies name, sex and population assignment of each sample
- `snp` file - specifies positions of SNPs, REF/ALT alleles etc.
- `geno` file - specifies SNP data (one row per site) in a dense string-based format:
  - 0: individual is homozygous ALT
  - 1: individual is a heterozygote
  - 2: individual is homozygous REF
  - 9: missing data
  
As you can see, a VCF file is essentially a combination of all three 
files in a single file. Luckily for us, all three EIGENSTRAT files 
usually share a common path and prefix (at least you should try to
make it so whenever you work with them). This allows us to work with
just the prefix, instead of worrying about individual files.

As such, all main `admixr` functions accept a `prefix` argument, 
which specifies the path and prefix of all three EIGENSTRAT files
(you can still work with individual files if you need to, using `ind`,
`snp` and `geno` arguments of each `admixr` function, but try to avoid
that, as it makes your code mode verbose).

Here is a prefix of a small testing SNP data set that's distributed with
`admixr`. We will be using this data set in the rest of this vignette.

```{r}
(eigenstrat <- file.path(system.file(package = "admixr", "extdata"), "snps"))
```

We can verify that there are three files with this prefix, as they should
be:

```{r}
dir(path = dirname(eigenstrat), full.names = TRUE)
```

Let's look at their contents.

#### `ind` file
```{r, echo = FALSE}
cat(system(paste0("cat ", eigenstrat, ".ind"), intern = TRUE), sep = "\n")
```

The first column (individual ID) and the third column (population label) are
generally not the same (individual IDs often having numerical suffixes, etc.),
but are the same here for simplicity. Importantly, when specifying population/sample
arguments in `admixr` functions, the information in the third column is what is
used. For example, if you have individuals such as "French1", "French2", "French3"
in the first column, all three sharing a "French" population label in the third
column, specifying "French" in an `admixr` command will combine all three samples
in a single population, and will calculate allele frequency from all of them.

#### `snp` file (first 3 lines)
```{r, echo = FALSE}
cat(system(paste0("head -n 3 ", eigenstrat, ".snp"), intern = TRUE), sep = "\n")
```

#### `geno` file (first 3 lines)
```{r, echo = FALSE}
cat(system(paste0("head -n 3 ", eigenstrat, ".geno"), intern = TRUE), sep = "\n")
```










## Philosophy of `admixr`

The goal of `admixr` is to make ADMIXTOOLS analyses as trivial to perform
as possible, without having to worry about par/pop/left/right
configuration files (as they are known in ADMIXTOOLS' jargon) and other
low-level details.

The only interface between you and ADMIXTOOLS is the following set of R functions:

- `d()`
- `f4()`
- `f4ratio()`
- `f3()`
- `qpAdm()`

Anything that would normally require [dozens of lines of shell scripts](https://gaworkshop.readthedocs.io/en/latest/contents/06_f3/f3.html)
can be (most of the time) accomplished by running a single line of R 
code.









The following sections describe the usage of `admixr` on a set of
example analyses that one might be interested in doing.








## $D$ statistic

Let's say we are interested in the following question:
_"Which populations today show evidence of Neanderthal admixture?_

One way of looking at this is using the following D statistic:
$$D(\textrm{present-day human W}, \textrm{African}, \textrm{Neanderthal}, \textrm{Chimp}).$$
All $D$ statistics are based on comparing the proportions of BABA and ABBA sites
patterns observed in data:

$$ D = \frac{\textrm{# BABA sites - # ABBA sites}}{\textrm{# BABA sites + # ABBA sites}}.$$

Significant departure of $D$ from zero indicates an excess of allele
sharing between the first and the third population (positive $D$),
or an excess of allele sharing between the second and the third population
(negative $D$). If we get $D$ that is not significantly different from 0,
this suggests that the first and second populations form a clade, and don't
differ in their genetic affinity to the third population.

Therefore, our $D$ statistic above simply tests whether some modern humans today
admixed with Neanderthals, which would increase their genetic affinity to this
archaic group compared to West Africans (whose ancestors never met Neanderthals).

Let's save the population names first to make the code below more readable:
```{r}
pops <- c("French", "Sardinian", "Han", "Papuan", "Khomani_San", "Mbuti", "Dinka")
```

Then we can calculate the $D$ statistic above by running:

```{r}
result <- d(W = pops, X = "Yoruba", Y = "Vindija", Z = "Chimp", prefix = eigenstrat)
```

Which will return the following `data.frame`:
```{r eval = FALSE}
head(result)
```

```{r, echo = FALSE}
knitr::kable(head(result))
```

We can see that this `data.frame` object contains all the input 
information, but contains additional columns:

- `D` - $D$ statistic value
- `stderr` - standard error of the $D$ statistic from the block jackknife
- `Zscore` - $Z$ significance value
- `BABA`/`ABBA` - counts of observed site patterns
- `nsnps` - number of SNPs used for the calculation in this row

The format of output tables from other other `admixr` functions is very 
similar. 

While we could certainly make some inferences straight from this table by looking
at the $Z$ scores, tables in general are not the best representation of this
kind of data, especially as the number of samples increases. This is how we can
use the [`ggplot2`](https://ggplot2.tidyverse.org) package to plot the results:

```{r, fig.width = 7, fig.height = 4}
ggplot(result, aes(fct_reorder(W, D), D, color = abs(Zscore) > 2)) +
  geom_point() +
  geom_errorbar(aes(ymin = D - 2 * stderr, ymax = D + 2 * stderr))
```

We can see that all three Africans have $D$ consistent with 0, meaning that
the data is consistent with the null hypothesis of no significant Neanderthal
ancestry in Africans. On the other hand, the test rejects the null hypothesis for
all non-Africans today, suggesting that Neanderthals admixed with the ancestors
of present-day non-Africans. In fact, this is a similar test to the one that was
used as evidence supporting the Neanderthal admixture hypothesis in the first place!








## $f_4$ statistic

An alternative way of addressing the previous question is to use the 
$f_4$ statistic, which is very similar to $D$ statistic and can be calculated as:

$$ f_4 = \frac{\textrm{# BABA sites - # ABBA sites}}{\textrm{# sites}}$$
Again, significant departure of $f_4$ from 0 is informative about gene flow,
in an analogous way to $D$ statistic.

To repeat the previous analysis using $f_4$ statistic, we can run:

```{r}
result <- f4(W = pops, X = "Yoruba", Y = "Vindija", Z = "Chimp", prefix = eigenstrat)
```

```{r eval = FALSE}
head(result)
```

```{r, echo = FALSE}
knitr::kable(head(result))
```

```{r, fig.width = 7, fig.height = 4}
ggplot(result, aes(fct_reorder(W, f4), f4, color = abs(Zscore) > 2)) +
  geom_point() +
  geom_errorbar(aes(ymin = f4 - 2 * stderr, ymax = f4 + 2 * stderr))
```

As we can see by comparing this result to the $D$ statistic result, we can make
the same conclusions.

You might be wondering why we have both $f_4$ and $D$ if they are so similar. The
truth is that $f_4$ is directly informative about the amount of shared genetic
drift (the "branch length") between pairs of populations which is, in many cases,
a very useful theoretical property. Other than that, it's often a matter of personal
preference, and so `admixr` provides a separate functions for calculating both.







## $f_4$-ratio statistic

Now we know that non-Africans today carry _some_ Neanderthal ancestry. But what if
we want to know _how much_ Neanderthal ancestry they have? What proportion of their
genomes is of Neanderthal origin?

In general, when we are interested in estimating the *proportion* of ancestry
coming from a parental lineage, we can use ratio of $f_4$ statistics. The theory
between $f_4$-ratios is slightly more complicated and we can't explain it here,
but if you're interested in the technical details I recommend Benjamin Peter's
[overview](http://www.genetics.org/content/early/2016/02/03/genetics.115.183913)
and Nick Patterson's [original ADMIXTOOLS paper](http://www.genetics.org/content/192/3/1065).

Using the nomenclature of Patterson et al. 2012, we can perform calculate $f_4$-ratios
using the following code (`X` being a vector of samples in which we want 
to estimate Neanderthal ancestry):

```{r}
result <- f4ratio(X = pops, A = "Altai", B = "Vindija", C = "Yoruba", O = "Chimp", prefix = eigenstrat)
```

The ancestry proportion (a number between 0 and 1) is given in the `alpha`
column:

```{r, eval=FALSE}
head(result)
```

```{r, echo=FALSE}
knitr::kable(head(result))
```

```{r, fig.width = 7, fig.height = 4}
ggplot(result, aes(fct_reorder(X, alpha), alpha, color = abs(Zscore) > 2)) +
  geom_point() +
  geom_errorbar(aes(ymin = alpha - 2 * stderr, ymax = alpha + 2 * stderr)) +
  geom_hline(yintercept = 0, linetype = 2) +
  labs(y = "Neandertal ancestry proportion", x = "present-day individual")
```

We can make several observations:

- Again, we don't see any significant Neanderthal ancestry in present-day 
  Africans (proportion consistent with 0%), which is what we confirmed using $D$
  and $f_4$ above.
- Present-day non-Africans carry between 2-3% of Neanderthal ancestry.
- We see a much higher proportion of Neanderthal ancestry in people from Papua New   
  Guinea - more than 4%!








## $f_3$ statistic

The $f_3$ statistic, also known as the 3-population statistic, is useful whenever we
want to:

1. Estimate the branch length (shared genetic drift) between a pair of populations
   $A$ and $B$ with respect to a common outgroup $C$. In this case, the higher the
   $f_3$ value, the longer the shared evolutionary time between $A$ and $B$.
2. Test whether population $C$ is a mixture of two parental populations $A$ and $B$.
   Negative value of the $f_3$ statistic then serves as statistical evidence of
   this admixture.

As an example, imagine we are interested in relative divergence times between pairs
of present-day human populations, and want to know in which approximate order they
split of from each other. To address this problem, we could use $f_3$ statistic by
fixing the $C$ outgroup as Chimp, and calculating pairwise $f_3$ statistics
between all pairs of present-day modern humans.


```{r}
pops <- c("French", "Sardinian", "Han", "Papuan", "Khomani_San", "Mbuti", "Dinka", "Yoruba")

result <- f3(A = pops, B = pops, C = "Chimp", prefix = eigenstrat)
```

```{r, eval=FALSE}
head(result)
```

```{r, echo=FALSE}
knitr::kable(head(result))
```


```{r, fig.width = 7, fig.height = 6}
# sort the population labels according to an increasing f3 value relative to French
ordered <- c(filter(result, A == "French") %>% arrange(f3) %>% .[["B"]], "French")

# plot heatmap of pairwise f3 values
result %>%
  mutate(A = factor(A, levels = ordered),
         B = factor(B, levels = ordered)) %>%
  ggplot(aes(A, B)) + geom_tile(aes(fill = f3))
```

We can see that when we order the heat map labels based on values of pairwise
$f_3$ statistics, the decreasing divergence times between human populations
(left to right and bottom up) shows beautifully.







## qpAdm method

The last ADMIXTOOLS method implemented in `admixr` is qpAdm. Unfortunately, it
is also one that is the most complex and has not been properly described and
peer-reviewed yet. Nevertheless, it seems to have lot of power to disentangle complex
admixture scenarios, and so we included it in our package as well.

Very briefly, qpAdm can be used to estimate admixture proportions coming from
a series of $N$ source *ancestral* populations, assuming we have reference populations 
that form clades with those *source* populations and are closer to them than to any of
the specified *outgroup* populations.

Formally, if the Test population has ancestry coming from $N$ ancestral source
populations, with Reference populations being closer to them than are outgroup
populations $O_i$, we can write:

$$f_4(\textrm{Test}, O_a, O_b, O_c) \approx \sum_{i=1}^N \alpha_i f_4(\textrm{Reference}_i, O_a; O_b, O_c),$$

where $\sum_{i=1}^N \alpha_i = 1$ and $\alpha_i \geq 1$ for all $i = 1, ..., N$.

If this looks like black magic to you, I sincerely apologize and direct you
to the Supplementary Section 9 of [Haak et al. 2015](http://www.nature.com/articles/nature14317), and an [informal write-up](https://github.com/DReichLab/AdmixTools/blob/master/pdoc.pdf)
distributed with the ADMIXTOOLS software.

Probably the simplest possible case to show that qpAdm works is by returning
to the question of estimating Neanderthal ancestry proportions. Let's define:

- *target* populations to estimate ancestry proportions for - Europeans in this case
- *reference* populations - Vindija Neanderthal and an African (two potential sources of
  ancestries in Europeans today)
- *outgroup* populations - Chimp, Altai Neanderthal and Denisovan (which are all
  further from the true ancestral populations than the *reference* populations)

Then we can run qpAdm with:

```{r}
result <- qpAdm(
  target = c("French", "Sardinian", "Mbuti", "Dinka"),
  refs = c("Vindija", "Yoruba"),
  outgroups = c("Chimp", "Denisova", "Altai"),
  prefix = eigenstrat
)
```

And we get the result as a `data.frame` again:
```{r, eval=FALSE}
result
```

```{r, echo=FALSE}
knitr::kable(result)
```


Note that we get admixture proportions standard errors for each potential
source in columns.
