<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>admixr - tutorial • admixr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="admixr - tutorial">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">admixr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/overview.html">admixr - tutorial</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bodkan/admixr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>admixr - tutorial</h1>
                        <h4 class="author">Martin Petr</h4>
            
            <h4 class="date">2018-08-25</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/bodkan/admixr/blob/master/vignettes/overview.Rmd"><code>vignettes/overview.Rmd</code></a></small>
      <div class="hidden name"><code>overview.Rmd</code></div>

    </div>

    
    
<p>This vignette describes how to calculate various population admixture statistics (<span class="math inline">\(D\)</span>, <span class="math inline">\(f_4\)</span>, etc.) using the <code>admixr</code> package. It assumes that you already know a little bit about them, because we don’t have enough space to go into details of how and why they work. If you want to learn more and are not afraid of little bit of math, I highly recommend Benjamin Peter’s <a href="http://www.genetics.org/content/early/2016/02/03/genetics.115.183913">wonderful overview</a> of the subject and Nick Patterson’s <a href="http://www.genetics.org/content/192/3/1065">original ADMIXTOOLS paper</a>.</p>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p><a href="http://www.genetics.org/content/192/3/1065">ADMIXTOOLS</a> is a widely used software package for calculating admixture statistics and testing population admixture hypotheses. However, although powerful and comprehensive, it is not really user-friendly.</p>
<p>A typical ADMIXTOOLS workflow usually involves a combination of <code>sed</code>/<code>awk</code>/shell scripting and manual editing to create different configuration files. These are then passed as command-line arguments to one of ADMIXTOOLS’ commands, and control how to run a particular analysis. The results are then redirected to another file, and the user needs to extract values of interest from this file (which is full of redundant information), often using additional command-line utilities, or (worse) by manual copy-pasting. The processed results are then analysed in R, Excel or another program.</p>
<p>This workflow is very cumbersome, especially if one wants to explore many hypotheses involving different combinations of populations. Most importantly, however, it makes it difficult to follow good practices of reproducible science, as it is nearly impossible to construct reproducible automated “pipelines”.</p>
<p>This R package makes it possible to perform all stages of ADMIXTOOLS analyses entirely from within R. It provides a set of convenient functions that completely remove the need for “low level” configuration of individual ADMIXTOOLS programs, allowing users to focus on the analysis itself.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>To install <code>admixr</code> from GitHub you need to install the package <code>devtools</code> first. To do this, you can simply run (in R):</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">install.packages</span>(<span class="st">"devtools"</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"bodkan/admixr"</span>)</a></code></pre></div>
<p>Furthermore, if you want to follow the examples in this vignette, you will need the <a href="https://www.tidyverse.org">tidyverse</a> collection of packages for convenient data analysis, which you can install with:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">install.packages</span>(<span class="st">"tidyverse"</span>)</a></code></pre></div>
<p>When everything is ready, you can load both packages:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw">library</span>(admixr)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="kw">library</span>(tidyverse)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="co">#&gt; ── Attaching packages ─────────────────────────────────────────────────── tidyverse 1.2.1 ──</span></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="co">#&gt; ✔ ggplot2 3.0.0     ✔ purrr   0.2.5</span></a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="co">#&gt; ✔ tibble  1.4.2     ✔ dplyr   0.7.6</span></a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="co">#&gt; ✔ tidyr   0.8.1     ✔ stringr 1.3.1</span></a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="co">#&gt; ✔ readr   1.1.1     ✔ forcats 0.3.0</span></a>
<a class="sourceLine" id="cb3-8" data-line-number="8"><span class="co">#&gt; ── Conflicts ────────────────────────────────────────────────────── tidyverse_conflicts() ──</span></a>
<a class="sourceLine" id="cb3-9" data-line-number="9"><span class="co">#&gt; ✖ dplyr::filter() masks stats::filter()</span></a>
<a class="sourceLine" id="cb3-10" data-line-number="10"><span class="co">#&gt; ✖ dplyr::lag()    masks stats::lag()</span></a></code></pre></div>
<p><strong>Note that in order to run <code>admixr</code> analyses, you need a working installation of ADMIXTOOLS!</strong> Explaining how to compile it is beyond the scope of this document, but you can find help <a href="https://github.com/DReichLab/AdmixTools">here</a>. <strong>Furthermore, you need to make sure that R can find ADMIXTOOLS binaries on the <code>$PATH</code>.</strong> If this is not the case, running <code>library(admixr)</code> will show a warning message with instructions on how to fix this.</p>
</div>
<div id="a-note-about-eigenstrat-files" class="section level2">
<h2 class="hasAnchor">
<a href="#a-note-about-eigenstrat-files" class="anchor"></a>A note about EIGENSTRAT files</h2>
<p>If you have an EIGENSTRAT “triplet” of files ready, and just want to know how to calculate different admixture statistics, feel free to skip this section.</p>
<div id="eigenstrat-file-format" class="section level3">
<h3 class="hasAnchor">
<a href="#eigenstrat-file-format" class="anchor"></a>EIGENSTRAT file format</h3>
<p>ADMIXTOOLS software uses a peculiar set of genetic file formats, which seem strange if you are used to working with <a href="http://samtools.github.io/hts-specs/VCFv4.3.pdf">VCF files</a>. However, the basic idea remains the same - we want to store and access SNP data (REF/ALT alleles) of a set of individuals at a defined set of genomic positions.</p>
<p>EIGENSTRAT data sets always contain three kinds of files:</p>
<ul>
<li>
<code>ind</code> file - specifies name, sex and population assignment of each sample</li>
<li>
<code>snp</code> file - specifies positions of SNPs, REF/ALT alleles etc.</li>
<li>
<code>geno</code> file - specifies SNP data (one row per site) in a dense string-based format:
<ul>
<li>0: individual is homozygous ALT</li>
<li>1: individual is a heterozygote</li>
<li>2: individual is homozygous REF</li>
<li>9: missing data</li>
</ul>
</li>
</ul>
<p>As you can see, a VCF file is essentially a combination of all three files in a single file. Luckily for us, all three EIGENSTRAT files usually share a common path and prefix (at least you should try to make it so whenever you work with them). This allows us to work with just the prefix, instead of worrying about individual files.</p>
<p>As such, all main <code>admixr</code> functions accept a <code>prefix</code> argument, which specifies the path and prefix of all three EIGENSTRAT files (you can still work with individual files if you need to, using <code>ind</code>, <code>snp</code> and <code>geno</code> arguments of each <code>admixr</code> function, but try to avoid that, as it makes your code mode verbose).</p>
<p>Here is a prefix of a small testing SNP data set that’s distributed with <code>admixr</code>. We will be using this data set in the rest of this vignette.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">(eigenstrat &lt;-<span class="st"> </span><span class="kw">file.path</span>(<span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">"admixr"</span>, <span class="st">"extdata"</span>), <span class="st">"snps"</span>))</a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="co">#&gt; [1] "/Users/martin_petr/local/R_LIBS/admixr/extdata/snps"</span></a></code></pre></div>
<p>We can verify that there are three files with this prefix, as they should be:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">dir</span>(<span class="dt">path =</span> <span class="kw">dirname</span>(eigenstrat), <span class="dt">full.names =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="co">#&gt; [1] "/Users/martin_petr/local/R_LIBS/admixr/extdata/snps.geno"</span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="co">#&gt; [2] "/Users/martin_petr/local/R_LIBS/admixr/extdata/snps.ind" </span></a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="co">#&gt; [3] "/Users/martin_petr/local/R_LIBS/admixr/extdata/snps.snp"</span></a></code></pre></div>
<p>Let’s look at their contents.</p>
<div id="ind-file" class="section level4">
<h4 class="hasAnchor">
<a href="#ind-file" class="anchor"></a><code>ind</code> file</h4>
<pre><code>#&gt; Chimp    U   Chimp
#&gt; Mbuti    U   Mbuti
#&gt; Yoruba   U   Yoruba
#&gt; Dai  U   Dai
#&gt; Khomani_San  U   Khomani_San
#&gt; Han  U   Han
#&gt; Dinka    U   Dinka
#&gt; Sardinian    U   Sardinian
#&gt; Papuan   U   Papuan
#&gt; French   U   French
#&gt; Vindija  U   Vindija
#&gt; Altai    U   Altai
#&gt; Denisova U   Denisova</code></pre>
<p>The first column (individual ID) and the third column (population label) are generally not the same (individual IDs often having numerical suffixes, etc.), but are the same here for simplicity. Importantly, when specifying population/sample arguments in <code>admixr</code> functions, the information in the third column is what is used. For example, if you have individuals such as “French1”, “French2”, “French3” in the first column, all three sharing a “French” population label in the third column, specifying “French” in an <code>admixr</code> command will combine all three samples in a single population, and will calculate allele frequency from all of them.</p>
</div>
<div id="snp-file-first-3-lines" class="section level4">
<h4 class="hasAnchor">
<a href="#snp-file-first-3-lines" class="anchor"></a><code>snp</code> file (first 3 lines)</h4>
<pre><code>#&gt; 1_777122 1   0.007771    777122  A   T
#&gt; 1_832756 1   0.008328    832756  T   G
#&gt; 1_834832 1   0.008348    834832  G   C</code></pre>
</div>
<div id="geno-file-first-3-lines" class="section level4">
<h4 class="hasAnchor">
<a href="#geno-file-first-3-lines" class="anchor"></a><code>geno</code> file (first 3 lines)</h4>
<pre><code>#&gt; 9220000000000
#&gt; 9022021012000
#&gt; 9022021012000</code></pre>
</div>
</div>
</div>
<div id="philosophy-of-admixr" class="section level2">
<h2 class="hasAnchor">
<a href="#philosophy-of-admixr" class="anchor"></a>Philosophy of <code>admixr</code>
</h2>
<p>The goal of <code>admixr</code> is to make ADMIXTOOLS analyses as trivial to perform as possible, without having to worry about par/pop/left/right configuration files (as they are known in ADMIXTOOLS’ jargon) and other low-level details.</p>
<p>The only interface between you and ADMIXTOOLS is the following set of R functions:</p>
<ul>
<li><code><a href="../reference/d.html">d()</a></code></li>
<li><code><a href="../reference/f4.html">f4()</a></code></li>
<li><code><a href="../reference/f4ratio.html">f4ratio()</a></code></li>
<li><code><a href="../reference/f3.html">f3()</a></code></li>
<li><code><a href="../reference/qpAdm.html">qpAdm()</a></code></li>
</ul>
<p>Anything that would normally require <a href="https://gaworkshop.readthedocs.io/en/latest/contents/06_f3/f3.html">dozens of lines of shell scripts</a> can be (most of the time) accomplished by running a single line of R code.</p>
<p>The following sections describe the usage of <code>admixr</code> on a set of example analyses that one might be interested in doing.</p>
</div>
<div id="d-statistic" class="section level2">
<h2 class="hasAnchor">
<a href="#d-statistic" class="anchor"></a><span class="math inline">\(D\)</span> statistic</h2>
<p>Let’s say we are interested in the following question: <em>"Which populations today show evidence of Neanderthal admixture?</em></p>
<p>One way of looking at this is using the following D statistic: <span class="math display">\[D(\textrm{present-day human W}, \textrm{African}, \textrm{Neanderthal}, \textrm{Chimp}).\]</span> All <span class="math inline">\(D\)</span> statistics are based on comparing the proportions of BABA and ABBA sites patterns observed in data:</p>
<p><span class="math display">\[ D = \frac{\textrm{# BABA sites - # ABBA sites}}{\textrm{# BABA sites + # ABBA sites}}.\]</span></p>
<p>Significant departure of <span class="math inline">\(D\)</span> from zero indicates an excess of allele sharing between the first and the third population (positive <span class="math inline">\(D\)</span>), or an excess of allele sharing between the second and the third population (negative <span class="math inline">\(D\)</span>). If we get <span class="math inline">\(D\)</span> that is not significantly different from 0, this suggests that the first and second populations form a clade, and don’t differ in their genetic affinity to the third population.</p>
<p>Therefore, our <span class="math inline">\(D\)</span> statistic above simply tests whether some modern humans today admixed with Neanderthals, which would increase their genetic affinity to this archaic group compared to West Africans (whose ancestors never met Neanderthals).</p>
<p>Let’s save the population names first to make the code below more readable:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">pops &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"French"</span>, <span class="st">"Sardinian"</span>, <span class="st">"Han"</span>, <span class="st">"Papuan"</span>, <span class="st">"Khomani_San"</span>, <span class="st">"Mbuti"</span>, <span class="st">"Dinka"</span>)</a></code></pre></div>
<p>Then we can calculate the <span class="math inline">\(D\)</span> statistic above by running:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/d.html">d</a></span>(<span class="dt">W =</span> pops, <span class="dt">X =</span> <span class="st">"Yoruba"</span>, <span class="dt">Y =</span> <span class="st">"Vindija"</span>, <span class="dt">Z =</span> <span class="st">"Chimp"</span>, <span class="dt">prefix =</span> eigenstrat)</a></code></pre></div>
<p>Which will return the following <code>data.frame</code>:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw">head</span>(result)</a></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">W</th>
<th align="left">X</th>
<th align="left">Y</th>
<th align="left">Z</th>
<th align="right">D</th>
<th align="right">stderr</th>
<th align="right">Zscore</th>
<th align="right">BABA</th>
<th align="right">ABBA</th>
<th align="right">nsnps</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">French</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.0233</td>
<td align="right">0.008729</td>
<td align="right">2.665</td>
<td align="right">6296</td>
<td align="right">6010</td>
<td align="right">195040</td>
</tr>
<tr class="even">
<td align="left">Sardinian</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.0184</td>
<td align="right">0.008292</td>
<td align="right">2.216</td>
<td align="right">6278</td>
<td align="right">6052</td>
<td align="right">194981</td>
</tr>
<tr class="odd">
<td align="left">Han</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.0344</td>
<td align="right">0.008920</td>
<td align="right">3.853</td>
<td align="right">6396</td>
<td align="right">5971</td>
<td align="right">195120</td>
</tr>
<tr class="even">
<td align="left">Papuan</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.0474</td>
<td align="right">0.008791</td>
<td align="right">5.387</td>
<td align="right">6475</td>
<td align="right">5890</td>
<td align="right">195040</td>
</tr>
<tr class="odd">
<td align="left">Khomani_San</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.0074</td>
<td align="right">0.008169</td>
<td align="right">0.906</td>
<td align="right">6485</td>
<td align="right">6390</td>
<td align="right">194960</td>
</tr>
<tr class="even">
<td align="left">Mbuti</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.0042</td>
<td align="right">0.008142</td>
<td align="right">0.514</td>
<td align="right">6330</td>
<td align="right">6277</td>
<td align="right">194994</td>
</tr>
</tbody>
</table>
<p>We can see that this <code>data.frame</code> object contains all the input information, but contains additional columns:</p>
<ul>
<li>
<code>D</code> - <span class="math inline">\(D\)</span> statistic value</li>
<li>
<code>stderr</code> - standard error of the <span class="math inline">\(D\)</span> statistic from the block jackknife</li>
<li>
<code>Zscore</code> - <span class="math inline">\(Z\)</span> significance value</li>
<li>
<code>BABA</code>/<code>ABBA</code> - counts of observed site patterns</li>
<li>
<code>nsnps</code> - number of SNPs used for the calculation in this row</li>
</ul>
<p>The format of output tables from other other <code>admixr</code> functions is very similar.</p>
<p>While we could certainly make some inferences straight from this table by looking at the <span class="math inline">\(Z\)</span> scores, tables in general are not the best representation of this kind of data, especially as the number of samples increases. This is how we can use the <a href="https://ggplot2.tidyverse.org"><code>ggplot2</code></a> package to plot the results:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="kw">ggplot</span>(result, <span class="kw">aes</span>(<span class="kw">fct_reorder</span>(W, D), D, <span class="dt">color =</span> <span class="kw">abs</span>(Zscore) <span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb12-3" data-line-number="3"><span class="st">  </span><span class="kw">geom_errorbar</span>(<span class="kw">aes</span>(<span class="dt">ymin =</span> D <span class="op">-</span><span class="st"> </span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span>stderr, <span class="dt">ymax =</span> D <span class="op">+</span><span class="st"> </span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span>stderr))</a></code></pre></div>
<p><img src="overview_files/figure-html/unnamed-chunk-13-1.png" width="672"></p>
<p>We can see that all three Africans have <span class="math inline">\(D\)</span> consistent with 0, meaning that the data is consistent with the null hypothesis of no significant Neanderthal ancestry in Africans. On the other hand, the test rejects the null hypothesis for all non-Africans today, suggesting that Neanderthals admixed with the ancestors of present-day non-Africans. In fact, this is a similar test to the one that was used as evidence supporting the Neanderthal admixture hypothesis in the first place!</p>
</div>
<div id="f_4-statistic" class="section level2">
<h2 class="hasAnchor">
<a href="#f_4-statistic" class="anchor"></a><span class="math inline">\(f_4\)</span> statistic</h2>
<p>An alternative way of addressing the previous question is to use the <span class="math inline">\(f_4\)</span> statistic, which is very similar to <span class="math inline">\(D\)</span> statistic and can be calculated as:</p>
<p><span class="math display">\[ f_4 = \frac{\textrm{# BABA sites - # ABBA sites}}{\textrm{# sites}}\]</span> Again, significant departure of <span class="math inline">\(f_4\)</span> from 0 is informative about gene flow, in an analogous way to <span class="math inline">\(D\)</span> statistic.</p>
<p>To repeat the previous analysis using <span class="math inline">\(f_4\)</span> statistic, we can run:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/f4.html">f4</a></span>(<span class="dt">W =</span> pops, <span class="dt">X =</span> <span class="st">"Yoruba"</span>, <span class="dt">Y =</span> <span class="st">"Vindija"</span>, <span class="dt">Z =</span> <span class="st">"Chimp"</span>, <span class="dt">prefix =</span> eigenstrat)</a></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw">head</span>(result)</a></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">W</th>
<th align="left">X</th>
<th align="left">Y</th>
<th align="left">Z</th>
<th align="right">f4</th>
<th align="right">stderr</th>
<th align="right">Zscore</th>
<th align="right">BABA</th>
<th align="right">ABBA</th>
<th align="right">nsnps</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">French</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.001468</td>
<td align="right">0.000551</td>
<td align="right">2.663</td>
<td align="right">6296</td>
<td align="right">6010</td>
<td align="right">195040</td>
</tr>
<tr class="even">
<td align="left">Sardinian</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.001162</td>
<td align="right">0.000525</td>
<td align="right">2.212</td>
<td align="right">6278</td>
<td align="right">6052</td>
<td align="right">194981</td>
</tr>
<tr class="odd">
<td align="left">Han</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.002178</td>
<td align="right">0.000567</td>
<td align="right">3.838</td>
<td align="right">6396</td>
<td align="right">5971</td>
<td align="right">195120</td>
</tr>
<tr class="even">
<td align="left">Papuan</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.003002</td>
<td align="right">0.000559</td>
<td align="right">5.374</td>
<td align="right">6475</td>
<td align="right">5890</td>
<td align="right">195040</td>
</tr>
<tr class="odd">
<td align="left">Khomani_San</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.000489</td>
<td align="right">0.000540</td>
<td align="right">0.905</td>
<td align="right">6485</td>
<td align="right">6390</td>
<td align="right">194960</td>
</tr>
<tr class="even">
<td align="left">Mbuti</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.000271</td>
<td align="right">0.000526</td>
<td align="right">0.514</td>
<td align="right">6330</td>
<td align="right">6277</td>
<td align="right">194994</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="kw">ggplot</span>(result, <span class="kw">aes</span>(<span class="kw">fct_reorder</span>(W, f4), f4, <span class="dt">color =</span> <span class="kw">abs</span>(Zscore) <span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb15-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb15-3" data-line-number="3"><span class="st">  </span><span class="kw">geom_errorbar</span>(<span class="kw">aes</span>(<span class="dt">ymin =</span> f4 <span class="op">-</span><span class="st"> </span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span>stderr, <span class="dt">ymax =</span> f4 <span class="op">+</span><span class="st"> </span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span>stderr))</a></code></pre></div>
<p><img src="overview_files/figure-html/unnamed-chunk-17-1.png" width="672"></p>
<p>As we can see by comparing this result to the <span class="math inline">\(D\)</span> statistic result, we can make the same conclusions.</p>
<p>You might be wondering why we have both <span class="math inline">\(f_4\)</span> and <span class="math inline">\(D\)</span> if they are so similar. The truth is that <span class="math inline">\(f_4\)</span> is directly informative about the amount of shared genetic drift (the “branch length”) between pairs of populations which is, in many cases, a very useful theoretical property. Other than that, it’s often a matter of personal preference, and so <code>admixr</code> provides a separate functions for calculating both.</p>
</div>
<div id="f_4-ratio-statistic" class="section level2">
<h2 class="hasAnchor">
<a href="#f_4-ratio-statistic" class="anchor"></a><span class="math inline">\(f_4\)</span>-ratio statistic</h2>
<p>Now we know that non-Africans today carry <em>some</em> Neanderthal ancestry. But what if we want to know <em>how much</em> Neanderthal ancestry they have? What proportion of their genomes is of Neanderthal origin?</p>
<p>In general, when we are interested in estimating the <em>proportion</em> of ancestry coming from a parental lineage, we can use ratio of <span class="math inline">\(f_4\)</span> statistics. The theory between <span class="math inline">\(f_4\)</span>-ratios is slightly more complicated and we can’t explain it here, but if you’re interested in the technical details XXX</p>
<p>Using the nomenclature of Patterson et al. 2012, we can perform calculate <span class="math inline">\(f_4\)</span>-ratios using the following code (<code>X</code> being a vector of samples in which we want to estimate Neanderthal ancestry):</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/f4ratio.html">f4ratio</a></span>(<span class="dt">X =</span> pops, <span class="dt">A =</span> <span class="st">"Altai"</span>, <span class="dt">B =</span> <span class="st">"Vindija"</span>, <span class="dt">C =</span> <span class="st">"Yoruba"</span>, <span class="dt">O =</span> <span class="st">"Chimp"</span>, <span class="dt">prefix =</span> eigenstrat )</a></code></pre></div>
<p>The ancestry proportion (a number between 0 and 1) is given in the <code>alpha</code> column:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="kw">head</span>(result)</a></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">A</th>
<th align="left">B</th>
<th align="left">X</th>
<th align="left">C</th>
<th align="left">O</th>
<th align="right">alpha</th>
<th align="right">stderr</th>
<th align="right">Zscore</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">French</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.017916</td>
<td align="right">0.007773</td>
<td align="right">2.305</td>
</tr>
<tr class="even">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Sardinian</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.016491</td>
<td align="right">0.007376</td>
<td align="right">2.236</td>
</tr>
<tr class="odd">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Han</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.028865</td>
<td align="right">0.007855</td>
<td align="right">3.675</td>
</tr>
<tr class="even">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Papuan</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.041625</td>
<td align="right">0.007808</td>
<td align="right">5.331</td>
</tr>
<tr class="odd">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Khomani_San</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.004076</td>
<td align="right">0.007621</td>
<td align="right">0.535</td>
</tr>
<tr class="even">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Mbuti</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.004688</td>
<td align="right">0.007357</td>
<td align="right">0.637</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1"><span class="kw">ggplot</span>(result, <span class="kw">aes</span>(<span class="kw">fct_reorder</span>(X, alpha), alpha, <span class="dt">color =</span> <span class="kw">abs</span>(Zscore) <span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb18-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb18-3" data-line-number="3"><span class="st">  </span><span class="kw">geom_errorbar</span>(<span class="kw">aes</span>(<span class="dt">ymin =</span> alpha <span class="op">-</span><span class="st"> </span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span>stderr, <span class="dt">ymax =</span> alpha <span class="op">+</span><span class="st"> </span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span>stderr)) <span class="op">+</span></a>
<a class="sourceLine" id="cb18-4" data-line-number="4"><span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="dv">0</span>, <span class="dt">linetype =</span> <span class="dv">2</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb18-5" data-line-number="5"><span class="st">  </span><span class="kw">labs</span>(<span class="dt">y =</span> <span class="st">"Neandertal ancestry proportion"</span>, <span class="dt">x =</span> <span class="st">"present-day individual"</span>)</a></code></pre></div>
<p><img src="overview_files/figure-html/unnamed-chunk-21-1.png" width="672"></p>
<p>We can make several observations:</p>
<ul>
<li>Again, we don’t see any significant Neanderthal ancestry in present-day Africans (proportion consistent with 0%), which is what we confirmed using <span class="math inline">\(D\)</span> and <span class="math inline">\(f_4\)</span> above.</li>
<li>Present-day non-Africans carry between 2-3% of Neanderthal ancestry.</li>
<li>We see a much higher proportion of Neanderthal ancestry in people from Papua New<br>
Guinea - more than 4%!</li>
</ul>
</div>
<div id="f_3-statistic" class="section level2">
<h2 class="hasAnchor">
<a href="#f_3-statistic" class="anchor"></a><span class="math inline">\(f_3\)</span> statistic</h2>
<p>The <span class="math inline">\(f_3\)</span> statistic, also known as the 3-population statistic, is useful whenever we want to:</p>
<ol style="list-style-type: decimal">
<li>Estimate the branch length (shared genetic drift) between a pair of populations <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span> with respect to a common outgroup <span class="math inline">\(C\)</span>. In this case, the higher the <span class="math inline">\(f_3\)</span> value, the longer the shared evolutionary time between <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span>.</li>
<li>Test whether population <span class="math inline">\(C\)</span> is a mixture of two parental populations <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span>. Negative value of the <span class="math inline">\(f_3\)</span> statistic then serves as statistical evidence of this admixture.</li>
</ol>
<p>As an example, imagine we are interested in relative divergence times between pairs of present-day human populations, and want to know in which approximate order they split of from each other. To address this problem, we could use <span class="math inline">\(f_3\)</span> statistic by fixing the <span class="math inline">\(C\)</span> outgroup as Chimp, and calculating pairwise <span class="math inline">\(f_3\)</span> statistics between all pairs of present-day modern humans.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1">pops &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"French"</span>, <span class="st">"Sardinian"</span>, <span class="st">"Han"</span>, <span class="st">"Papuan"</span>, <span class="st">"Khomani_San"</span>, <span class="st">"Mbuti"</span>, <span class="st">"Dinka"</span>, <span class="st">"Yoruba"</span>)</a>
<a class="sourceLine" id="cb19-2" data-line-number="2"></a>
<a class="sourceLine" id="cb19-3" data-line-number="3">result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/f3.html">f3</a></span>(<span class="dt">A =</span> pops, <span class="dt">B =</span> pops, <span class="dt">C =</span> <span class="st">"Chimp"</span>, <span class="dt">prefix =</span> eigenstrat)</a></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="kw">head</span>(result)</a></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">A</th>
<th align="left">B</th>
<th align="left">C</th>
<th align="right">f3</th>
<th align="right">stderr</th>
<th align="right">Zscore</th>
<th align="right">nsnps</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">French</td>
<td align="left">Sardinian</td>
<td align="left">Chimp</td>
<td align="right">6.991000e+14</td>
<td align="right">2.035678e+12</td>
<td align="right">343.424</td>
<td align="right">90378</td>
</tr>
<tr class="even">
<td align="left">French</td>
<td align="left">Han</td>
<td align="left">Chimp</td>
<td align="right">6.658685e+14</td>
<td align="right">1.982191e+12</td>
<td align="right">335.925</td>
<td align="right">91942</td>
</tr>
<tr class="odd">
<td align="left">French</td>
<td align="left">Papuan</td>
<td align="left">Chimp</td>
<td align="right">6.532613e+14</td>
<td align="right">2.068070e+12</td>
<td align="right">315.880</td>
<td align="right">90734</td>
</tr>
<tr class="even">
<td align="left">French</td>
<td align="left">Khomani_San</td>
<td align="left">Chimp</td>
<td align="right">4.884753e+14</td>
<td align="right">1.665916e+12</td>
<td align="right">293.217</td>
<td align="right">97516</td>
</tr>
<tr class="odd">
<td align="left">French</td>
<td align="left">Mbuti</td>
<td align="left">Chimp</td>
<td align="right">5.061458e+14</td>
<td align="right">1.616864e+12</td>
<td align="right">313.042</td>
<td align="right">101399</td>
</tr>
<tr class="even">
<td align="left">French</td>
<td align="left">Dinka</td>
<td align="left">Chimp</td>
<td align="right">5.651877e+14</td>
<td align="right">1.722862e+12</td>
<td align="right">328.052</td>
<td align="right">103154</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1"><span class="co"># sort the population labels according to an increasing f3 value relative to French</span></a>
<a class="sourceLine" id="cb21-2" data-line-number="2">ordered &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">filter</span>(result, A <span class="op">==</span><span class="st"> "French"</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">arrange</span>(f3) <span class="op">%&gt;%</span><span class="st"> </span>.[[<span class="st">"B"</span>]], <span class="st">"French"</span>)</a>
<a class="sourceLine" id="cb21-3" data-line-number="3"></a>
<a class="sourceLine" id="cb21-4" data-line-number="4"><span class="co"># plot heatmap of pairwise f3 values</span></a>
<a class="sourceLine" id="cb21-5" data-line-number="5">result <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb21-6" data-line-number="6"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">A =</span> <span class="kw">factor</span>(A, <span class="dt">levels =</span> ordered),</a>
<a class="sourceLine" id="cb21-7" data-line-number="7">         <span class="dt">B =</span> <span class="kw">factor</span>(B, <span class="dt">levels =</span> ordered)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb21-8" data-line-number="8"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(A, B)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_tile</span>(<span class="kw">aes</span>(<span class="dt">fill =</span> f3)) <span class="op">+</span><span class="st"> </span><span class="kw">theme_classic</span>()</a></code></pre></div>
<p><img src="overview_files/figure-html/unnamed-chunk-25-1.png" width="672"></p>
<p>We can see that when we order the heat map labels based on values of pairwise <span class="math inline">\(f_3\)</span> statistics, the decreasing divergence times between human populations (left to right and bottom up) shows beautifully.</p>
</div>
<div id="qpadm-method" class="section level2">
<h2 class="hasAnchor">
<a href="#qpadm-method" class="anchor"></a>qpAdm method</h2>
<p>The last ADMIXTOOLS method implemented in <code>admixr</code> is qpAdm. Unfortunately, it is also one that is the most complex and has not been properly described and peer-reviewed yet. Nevertheless, it has a lot of power to disentangle complex admixture scenarios and we provide a convenient wrapper for it as well.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#installation">Installation</a></li>
      <li><a href="#a-note-about-eigenstrat-files">A note about EIGENSTRAT files</a></li>
      <li><a href="#philosophy-of-admixr">Philosophy of <code>admixr</code></a></li>
      <li><a href="#d-statistic"><span class="math inline">\(D\)</span> statistic</a></li>
      <li><a href="#f_4-statistic"><span class="math inline">\(f_4\)</span> statistic</a></li>
      <li><a href="#f_4-ratio-statistic"><span class="math inline">\(f_4\)</span>-ratio statistic</a></li>
      <li><a href="#f_3-statistic"><span class="math inline">\(f_3\)</span> statistic</a></li>
      <li><a href="#qpadm-method">qpAdm method</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
