<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>admixr - paper • admixr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="admixr - paper">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">admixr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/paper.html">admixr - paper</a>
    </li>
    <li>
      <a href="../articles/tutorial.html">admixr - tutorial</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bodkan/admixr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>admixr - paper</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/bodkan/admixr/blob/master/R/vignettes/paper.Rmd"><code>R/vignettes/paper.Rmd</code></a></small>
      <div class="hidden name"><code>paper.Rmd</code></div>

    </div>

    
    
<div id="admixr---r-package-for-facilitating-admixtools-analyses" class="section level1">
<h1 class="hasAnchor">
<a href="#admixr---r-package-for-facilitating-admixtools-analyses" class="anchor"></a>admixr - R package for facilitating ADMIXTOOLS analyses</h1>
<div id="abstract" class="section level2">
<h2 class="hasAnchor">
<a href="#abstract" class="anchor"></a>Abstract</h2>
<p>We present a new R package <em>admixr</em>, which provides a convenient interface for the ADMIXTOOLS software suite. In a traditional ADMIXTOOLS workflow, the user must first generate a set of text configuration files tailored to each individual analysis, often using a combination of shell scripting and manual editing. After executing an ADMIXTOOLS program on the command line using the provided configuration files, generated log files need to be parsed to extract values of interest before they can be used in a downstream analysis. Our package simplifies this process by automating all low-level configuration and parsing steps without any user intervention, making the previously complex workflow as simple as running a single line of R code. Furthermore, by unifying all steps of an analytic workflow under a single framework, this package allows building of fully automated analytic pipelines, significantly improving the reproducibility of population genetic studies.</p>
<p><strong>Source code and documentation:</strong> Source code, installation instructions, documentation and a tutorial can be found on the package website at <a href="http://bodkan.net/admixr" class="uri">http://bodkan.net/admixr</a>.</p>
</div>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>The growing numbers of available ancient and modern DNA samples have transformed our understanding of the evolutionary history of humans and other species. Several statistical methods have been developed to make inferences about past population movements and admixture events from genomic data <span class="citation">(Racimo <em>et al.</em>, 2015)</span>. Chief among those has been a series of admixture statistics such as <span class="math inline">\(f_3\)</span>, <span class="math inline">\(f_4\)</span>, <span class="math inline">\(f_4\)</span>-ratio and <span class="math inline">\(D\)</span> as implemented in the set of command-line utilities distributed in the ADMIXTOOLS software package <span class="citation">(Patterson <em>et al.</em>, 2012)</span>. Although instrumental in making inferences about ancient and modern human history - indeed, they have been used in nearly all recent studies of ancient DNA in high-profile scientific journals <span class="citation">(Fu <em>et al.</em>, 2016; Hajdinjak <em>et al.</em>, 2018; Lazaridis <em>et al.</em>, 2016)</span> - ADMIXTOOLS utilities suffer from several disadvantages. First, each individual analysis or hypothesis test rely on a set of text configuration files which have to be generated using a combination of shell scripting and manual editing. Furthermore, after running an ADMIXTOOLS command on the command-line, the user needs to extract relevant values from a log file before they can be imported in another software for further analysis and plotting. This workflow is quite cumbersome and error-prone, especially if the user needs to quickly iterate through complex hypotheses involving many different populations or samples. Most importantly, however, any reliance on manual user intervention for both the preparation and parsing of configuration and output log files makes it difficult to follow the rules of good practice of reproducible research. To overcome these challenges, we present a new R package which provides a unified interface for population admixture analyses, utilizing ADMIXTOOLS software suite for the underlying calculations. The package completely automates the generation, processing and parsing of all intermediate files, hiding all low-level details from the users and allowing them to focus on the analysis itself. Importantly, in addition to providing a more user-friendly interface, unifying the entire analytic workflow in an R environment makes it possible to implement fully automated reproducible analytic pipelines.</p>
</div>
<div id="implementation-and-example-usage" class="section level2">
<h2 class="hasAnchor">
<a href="#implementation-and-example-usage" class="anchor"></a>Implementation and example usage</h2>
<p>The <em>admixr</em> package is implemented using the R programming language <span class="citation">(R Core Team, 2018)</span>. It’s core functionality includes the following set of R functions: <code><a href="../reference/f4ratio.html">f3()</a></code>, <code><a href="../reference/f4ratio.html">d()</a></code>, <code><a href="../reference/f4ratio.html">f4()</a></code>, <code><a href="../reference/f4ratio.html">f4ratio()</a></code> and <code><a href="../reference/qpAdm.html">qpAdm()</a></code>, each corresponding to one of the population admixture statistics implemented in ADMIXTOOLS. Traditionally, performing even the most trivial analysis using the ADMIXTOOLS package presents a significant amount of overhead for the user. For example, in order to calculate <span class="math inline">\(f_4\)</span>-ratio statistic such as</p>
<p><span class="math display">\[f_4(\textrm{Altai}, \textrm{Chimp}; \textrm{X}, \textrm{Mbuti}) /
  f_4(\textrm{Altai}, \textrm{Chimp}; \textrm{Vindija}, \textrm{Mbuti}),\]</span></p>
<p>that is, to estimate the proportion of Neandertal ancestry in a set of individuals <span class="math inline">\(X\)</span>, the user needs to create parameter files and population list files, tailored for this specific analysis, run the <code>qpF4ratio</code> program on the command-line, and to parse the log file to obtain relevant information (see Supplementary Information for an example of performing the complete analysis using a traditional ADMIXTOOLS command-line-based workflow). Note that changing the analysis setup in a slightly different way or including a different set of samples requires re-generating all files again.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1">snp_data &lt;-<span class="st"> </span><span class="kw"><a href="../reference/eigenstrat.html">eigenstrat</a></span>(<span class="kw"><a href="../reference/download_data.html">download_data</a></span>())</a></code></pre></div>
<p>In contrast, the same analysis using the <em>admixr</em> package can be performed using just a single R command:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/f4ratio.html">f4ratio</a></span>(</a>
<a class="sourceLine" id="cb2-2" data-line-number="2">    <span class="dt">X =</span> <span class="kw">c</span>(<span class="st">"French"</span>, <span class="st">"Han"</span>, <span class="st">"Papuan"</span>, <span class="st">"Yoruba"</span>),</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">    <span class="dt">A =</span> <span class="st">"Altai"</span>, <span class="dt">B =</span> <span class="st">"Vindija"</span>, <span class="dt">C =</span> <span class="st">"Mbuti"</span>, <span class="dt">O =</span> <span class="st">"Chimp"</span>,</a>
<a class="sourceLine" id="cb2-4" data-line-number="4">    <span class="dt">data =</span> snp_data</a>
<a class="sourceLine" id="cb2-5" data-line-number="5">)</a></code></pre></div>
<p>Internally, this code generates all necessary configuration files, runs the <code>qpF4ratio</code> command, parses the output file, and presents the user with the following R <code>data.frame</code> object for further statistical analysis an plotting:</p>
<table class="table">
<thead><tr class="header">
<th align="left">A</th>
<th align="left">B</th>
<th align="left">X</th>
<th align="left">C</th>
<th align="left">O</th>
<th align="right">alpha</th>
<th align="right">stderr</th>
<th align="right">Zscore</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">French</td>
<td align="left">Mbuti</td>
<td align="left">Chimp</td>
<td align="right">0.023339</td>
<td align="right">0.006153</td>
<td align="right">3.793</td>
</tr>
<tr class="even">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Han</td>
<td align="left">Mbuti</td>
<td align="left">Chimp</td>
<td align="right">0.021510</td>
<td align="right">0.006232</td>
<td align="right">3.451</td>
</tr>
<tr class="odd">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Papuan</td>
<td align="left">Mbuti</td>
<td align="left">Chimp</td>
<td align="right">0.036785</td>
<td align="right">0.006110</td>
<td align="right">6.020</td>
</tr>
<tr class="even">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Yoruba</td>
<td align="left">Mbuti</td>
<td align="left">Chimp</td>
<td align="right">-0.000285</td>
<td align="right">0.005724</td>
<td align="right">-0.050</td>
</tr>
</tbody>
</table>
<p>Other ADMIXTOOLS analyses can be performed in a similar way, and are described in the tutorial R vignette in a more detail.</p>
</div>
<div id="additional-functionality" class="section level2">
<h2 class="hasAnchor">
<a href="#additional-functionality" class="anchor"></a>Additional functionality</h2>
<p>The fact that ADMIXTOOLS requires data to be in a peculiar EIGENSTRAT file format makes data processing and filtering can be challenging because this format is not supported by most of the standard bioinformatic tools. To simplify handling and processing of EIGENSTRAT datasets, <em>admixr</em> implements a set of complementary functions that can be useful during data-preparation:</p>
<ul>
<li>Filtering of SNP data based on a given BED file — <code>filter_sites()</code>.</li>
<li>Filtering out potential ancient DNA damage substitutions — <code>filter_damage()</code>.</li>
<li>Merging of EIGENSTRAT datasets — <code><a href="../reference/merge_eigenstrat.html">merge_eigenstrat()</a></code>.</li>
<li>Calculating coverage of sites in each sample — <code><a href="../reference/count_snps.html">count_snps()</a></code>.</li>
<li>Merging of individual population labels into larger groups — <code>group_labels()</code>.</li>
<li>VCF-EIGENSTRAT format conversion — <code><a href="../reference/vcf_to_eigenstrat.html">vcf_to_eigenstrat()</a></code>, <code><a href="../reference/vcf_to_eigenstrat.html">eigenstrat_to_vcf()</a></code>.</li>
</ul>
</div>
<div id="funding" class="section level2">
<h2 class="hasAnchor">
<a href="#funding" class="anchor"></a>Funding</h2>
<p>Development of this software was supported by the Max Planck Society.</p>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-fu_genetic_2016">
<p>Fu,Q. <em>et al.</em> (2016) The genetic history of Ice Age Europe. <em>Nature</em>, <strong>534</strong>, 200–205.</p>
</div>
<div id="ref-hajdinjak_reconstructing_2018">
<p>Hajdinjak,M. <em>et al.</em> (2018) Reconstructing the genetic history of late Neanderthals. <em>Nature</em>, <strong>555</strong>, 652–656.</p>
</div>
<div id="ref-lazaridis_genomic_2016">
<p>Lazaridis,I. <em>et al.</em> (2016) Genomic insights into the origin of farming in the ancient Near East. <em>Nature</em>, <strong>536</strong>, 419–424.</p>
</div>
<div id="ref-patterson_ancient_2012">
<p>Patterson,N. <em>et al.</em> (2012) Ancient Admixture in Human History. <em>Genetics</em>, <strong>192</strong>, 1065–1093.</p>
</div>
<div id="ref-racimo_evidence_2015">
<p>Racimo,F. <em>et al.</em> (2015) Evidence for archaic adaptive introgression in humans. <em>Nature Reviews Genetics</em>, <strong>16</strong>, 359–371.</p>
</div>
<div id="ref-r_core_2018">
<p>R Core Team (2018) R: A language and environment for statistical computing R Foundation for Statistical Computing, Vienna, Austria.</p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li>
<a href="#admixr---r-package-for-facilitating-admixtools-analyses">admixr - R package for facilitating ADMIXTOOLS analyses</a><ul class="nav nav-pills nav-stacked">
<li><a href="#abstract">Abstract</a></li>
      <li><a href="#introduction">Introduction</a></li>
      <li><a href="#implementation-and-example-usage">Implementation and example usage</a></li>
      <li><a href="#additional-functionality">Additional functionality</a></li>
      <li><a href="#funding">Funding</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
