<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>admixr - tutorial • admixr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="admixr - tutorial">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">admixr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/paper.html">admixr - paper</a>
    </li>
    <li>
      <a href="../articles/tutorial.html">admixr - tutorial</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bodkan/admixr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>admixr - tutorial</h1>
                        <h4 class="author">Martin Petr</h4>
            
            <h4 class="date">2018-09-05</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/bodkan/admixr/blob/master/vignettes/tutorial.Rmd"><code>vignettes/tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>tutorial.Rmd</code></div>

    </div>

    
    
<p>This vignette describes how to calculate various population admixture statistics (<span class="math inline">\(D\)</span>, <span class="math inline">\(f_4\)</span>, etc.) using the <code>admixr</code> package, and how to use it to do simple filtering and processing of EIGENSTRAT datasets.</p>
<p><strong>A friendly warning</strong>: many of the statistics implemented here can be quite sensitive to assumptions about population histories or to errors in the data (causiing spurious correlations between similarly processed samples, especially in case of ancient DNA). If you want to use these methods, you should really understand the theory behind them first! I highly recommend reading Benjamin Peter’s <a href="http://www.genetics.org/content/early/2016/02/03/genetics.115.183913">wonderful overview</a> of the subject and Nick Patterson’s <a href="http://www.genetics.org/content/192/3/1065">original ADMIXTOOLS paper</a>.</p>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p><a href="https://github.com/DReichLab/AdmixTools/">ADMIXTOOLS</a> is a widely used software package for calculating admixture statistics and testing population admixture hypotheses. However, although powerful and comprehensive, it is not exactly known for being user-friendly.</p>
<p>A typical ADMIXTOOLS workflow often involves a combination of <code>sed</code>/<code>awk</code>/shell scripting and manual editing to create different configuration files. These are then passed as command-line arguments to one of ADMIXTOOLS commands, and control how to run a particular analysis. The results are then redirected to another file, which has to be parsed by the user to extract values of interest, often using command-line utilities again or (worse) by manual copy-pasting. Finally, the processed results are analysed in R, Excel or another program.</p>
<p>This workflow is very cumbersome, especially if one wants to explore many hypotheses involving different combinations of populations. Most importantly, however, it makes it difficult to follow the rules of best practice for reproducible science, as it is nearly impossible to construct reproducible automated “pipelines”.</p>
<p>This R package makes it possible to perform all stages of an ADMIXTOOLS analysis entirely from R. It provides a set of convenient functions that completely remove the need for “low level” configuration of individual ADMIXTOOLS programs, allowing users to focus on the analysis itself.</p>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p><strong>Note that in order to use the <code>admixr</code> package, you need a working installation of ADMIXTOOLS!</strong> You can find installation instructions <a href="https://github.com/DReichLab/AdmixTools/blob/master/README.INSTALL">here</a>.</p>
<p><strong>Furthermore, you need to make sure that R can find ADMIXTOOLS binaries on the <code>$PATH</code>.</strong> If this is not the case, running <code>library(admixr)</code> will show a warning message with instructions on how to fix this.</p>
<p>To install <code>admixr</code> from GitHub you need to install the package <code>devtools</code> first. To do this, you can simply run (in R):</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">install.packages</span>(<span class="st">"devtools"</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2">devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"bodkan/admixr"</span>)</a></code></pre></div>
<p>Furthermore, if you want to follow the examples in this vignette, you will need the <a href="https://www.tidyverse.org">tidyverse</a> collection of packages for convenient data analysis, which you can install with:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">install.packages</span>(<span class="st">"tidyverse"</span>)</a></code></pre></div>
<p>When everything is ready, you can run the following code to make functions in both packages available:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw">library</span>(admixr)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="kw">library</span>(tidyverse)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="co">#&gt; ── Attaching packages ────────────────────────────────── tidyverse 1.2.1 ──</span></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="co">#&gt; ✔ ggplot2 3.0.0     ✔ purrr   0.2.5</span></a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="co">#&gt; ✔ tibble  1.4.2     ✔ dplyr   0.7.6</span></a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="co">#&gt; ✔ tidyr   0.8.1     ✔ stringr 1.3.1</span></a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="co">#&gt; ✔ readr   1.1.1     ✔ forcats 0.3.0</span></a>
<a class="sourceLine" id="cb3-8" data-line-number="8"><span class="co">#&gt; ── Conflicts ───────────────────────────────────── tidyverse_conflicts() ──</span></a>
<a class="sourceLine" id="cb3-9" data-line-number="9"><span class="co">#&gt; ✖ dplyr::filter() masks stats::filter()</span></a>
<a class="sourceLine" id="cb3-10" data-line-number="10"><span class="co">#&gt; ✖ dplyr::lag()    masks stats::lag()</span></a></code></pre></div>
</div>
<div id="a-note-about-eigenstrat-format" class="section level2">
<h2 class="hasAnchor">
<a href="#a-note-about-eigenstrat-format" class="anchor"></a>A note about EIGENSTRAT format</h2>
<p>ADMIXTOOLS software uses a peculiar set of genetic file formats, which may seem strange if you are used to working with <a href="http://samtools.github.io/hts-specs/VCFv4.3.pdf">VCF files</a>. However, the basic idea remains the same - we want to store and access SNP data (REF/ALT alleles) of a set of individuals at a defined set of genomic positions.</p>
<p>EIGENSTRAT datasets always contain three kinds of files:</p>
<ul>
<li>
<code>ind</code> file - specifies the name, sex and population assignment of each sample</li>
<li>
<code>snp</code> file - specifies the positions of SNPs, REF/ALT alleles etc.</li>
<li>
<code>geno</code> file - contains SNP data (one row per site, one columne per sample) in a dense string-based format:
<ul>
<li>0: individual is homozygous ALT</li>
<li>1: individual is a heterozygote</li>
<li>2: individual is homozygous REF</li>
<li>9: missing data</li>
</ul>
</li>
</ul>
<p>As you can see, a VCF file is essentially a combination of all three files in a single file. Luckily for us, all three EIGENSTRAT files usually share a common path and prefix (at least you should try to make it so whenever you work with them). This allows us to work with just the prefix instead of worrying about individual files for every analysis.</p>
<p>As such, all main <code>admixr</code> functions accept a <code>prefix</code> argument, which specifies the path and prefix of all three EIGENSTRAT files (you can still work with individual files if you need to - using <code>ind</code>, <code>snp</code> and <code>geno</code> arguments of each <code>admixr</code> function - but try to avoid that because it makes your code mode verbose).</p>
<p>Let’s first download a small testing SNP dataset using a built-in <code>admixr</code> function. We will be using this dataset in the rest of this vignette. For our convenience, the function returns a prefix of the whole dataset for future reference. (The function downloads the data into a temporary directory, feel free to change the final destination using it’s <code>dirname</code> parameter).</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">(eigenstrat_prefix &lt;-<span class="st"> </span><span class="kw"><a href="../reference/download_data.html">download_data</a></span>(<span class="dt">dirname =</span> <span class="kw">tempdir</span>()))</a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="co">#&gt; [1] "/var/folders/kk/s4cwdkx90pscz314mp0hhz480000gn/T//Rtmpo5bR7i/snps/snps"</span></a></code></pre></div>
<p>We can verify that there are three files with this prefix:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw">dir</span>(<span class="dt">path =</span> <span class="kw">dirname</span>(eigenstrat_prefix), <span class="dt">full.names =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="co">#&gt; [1] "/var/folders/kk/s4cwdkx90pscz314mp0hhz480000gn/T//Rtmpo5bR7i/snps/snps.geno"</span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="co">#&gt; [2] "/var/folders/kk/s4cwdkx90pscz314mp0hhz480000gn/T//Rtmpo5bR7i/snps/snps.ind" </span></a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="co">#&gt; [3] "/var/folders/kk/s4cwdkx90pscz314mp0hhz480000gn/T//Rtmpo5bR7i/snps/snps.snp"</span></a></code></pre></div>
<p>Let’s look at their contents:</p>
<div id="ind-file" class="section level4">
<h4 class="hasAnchor">
<a href="#ind-file" class="anchor"></a><code>ind</code> file</h4>
<pre><code>#&gt; Chimp        U  Chimp
#&gt; Mbuti        U  Mbuti
#&gt; Yoruba       U  Yoruba
#&gt; Khomani_San  U  Khomani_San
#&gt; Han          U  Han
#&gt; Dinka        U  Dinka
#&gt; Sardinian    U  Sardinian
#&gt; Papuan       U  Papuan
#&gt; French       U  French
#&gt; Vindija      U  Vindija
#&gt; Altai        U  Altai
#&gt; Denisova     U  Denisova</code></pre>
<p>The first column (sample name) and the third column (population label) are generally not the same (sample names often have numerical suffixes, etc.), but we kept them the same for simplicity. Importantly, when specifying population/sample arguments in <code>admixr</code> functions, the information in the third column is what is used. For example, if you have individuals such as “French1”, “French2”, “French3” in the first column of an <code>ind</code> file, all three sharing a “French” population label in the third column, specifying “French” in an <code>admixr</code> function will combine all three samples in a single population and work with it as a whole.</p>
</div>
<div id="snp-file-first-3-lines" class="section level4">
<h4 class="hasAnchor">
<a href="#snp-file-first-3-lines" class="anchor"></a><code>snp</code> file (first 3 lines)</h4>
<pre><code>#&gt; 1_832756 1   0.008328    832756  T   G
#&gt; 1_838931 1   0.008389    838931  A   C
#&gt; 1_843249 1   0.008432    843249  A   T</code></pre>
</div>
<div id="geno-file-first-3-lines" class="section level4">
<h4 class="hasAnchor">
<a href="#geno-file-first-3-lines" class="anchor"></a><code>geno</code> file (first 3 lines)</h4>
<pre><code>#&gt; 902021012000
#&gt; 922221211222
#&gt; 922222122222</code></pre>
</div>
</div>
<div id="philosophy-of-admixr" class="section level2">
<h2 class="hasAnchor">
<a href="#philosophy-of-admixr" class="anchor"></a>Philosophy of <code>admixr</code>
</h2>
<p>The goal of <code>admixr</code> is to make ADMIXTOOLS analyses as trivial to perform as possible, without having to worry about par/pop/left/right configuration files (as they are known in ADMIXTOOLS’ jargon) and other low-level details.</p>
<p>The only interface between you and ADMIXTOOLS is the following set of R functions:</p>
<ul>
<li><code><a href="../reference/f4ratio.html">d()</a></code></li>
<li><code><a href="../reference/f4ratio.html">f4()</a></code></li>
<li><code><a href="../reference/f4ratio.html">f4ratio()</a></code></li>
<li><code><a href="../reference/f4ratio.html">f3()</a></code></li>
<li><code><a href="../reference/qpAdm.html">qpAdm()</a></code></li>
</ul>
<p>Anything that would normally require <a href="https://gaworkshop.readthedocs.io/en/latest/contents/06_f3/f3.html">dozens of lines of shell scripts</a> can be often accomplished by running a single line of R code.</p>
<p>The following couple of sections describe the usage of <code>admixr</code> on a set of example analyses that one might be interested in doing.</p>
</div>
<div id="d-statistic" class="section level2">
<h2 class="hasAnchor">
<a href="#d-statistic" class="anchor"></a><span class="math inline">\(D\)</span> statistic</h2>
<p>Let’s say we are interested in the following question: <em>"Which populations today show evidence of Neanderthal admixture?</em></p>
<p>One way of looking at this is using the following D statistic: <span class="math display">\[D(\textrm{present-day human W}, \textrm{African}, \textrm{Neanderthal}, \textrm{Chimp}).\]</span></p>
<p><span class="math inline">\(D\)</span> statistics are based on comparing the proportions of BABA and ABBA sites patterns observed in data:</p>
<p><span class="math display">\[D = \frac{\textrm{# BABA sites - # ABBA sites}}{\textrm{# BABA sites + # ABBA sites}}.\]</span></p>
<p>Significant departure of <span class="math inline">\(D\)</span> from zero indicates an excess of allele sharing between the first and the third population (positive <span class="math inline">\(D\)</span>), or an excess of allele sharing between the second and the third population (negative <span class="math inline">\(D\)</span>). If we get <span class="math inline">\(D\)</span> that is not significantly different from 0, this suggests that the first and second populations form a clade, and don’t differ in their genetic affinity to the third population (this is the null hypothesis that the data is compared against).</p>
<p>Therefore, our <span class="math inline">\(D\)</span> statistic above simply tests whether some modern humans today admixed with Neanderthals, which would increase their genetic affinity to this archaic group compared to West Africans (whose ancestors never met Neanderthals).</p>
<p>Let’s save the population names first to make the code below more readable:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">pops &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"French"</span>, <span class="st">"Sardinian"</span>, <span class="st">"Han"</span>, <span class="st">"Papuan"</span>, <span class="st">"Khomani_San"</span>, <span class="st">"Mbuti"</span>, <span class="st">"Dinka"</span>)</a></code></pre></div>
<p>Using the <code>admixr</code> package we can then calculate the <span class="math inline">\(D\)</span> statistic above simply by running:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/f4ratio.html">d</a></span>(<span class="dt">W =</span> pops, <span class="dt">X =</span> <span class="st">"Yoruba"</span>, <span class="dt">Y =</span> <span class="st">"Vindija"</span>, <span class="dt">Z =</span> <span class="st">"Chimp"</span>, <span class="dt">prefix =</span> eigenstrat_prefix)</a></code></pre></div>
<p>The result is a following <code>data.frame</code>:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw">head</span>(result)</a></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">W</th>
<th align="left">X</th>
<th align="left">Y</th>
<th align="left">Z</th>
<th align="right">D</th>
<th align="right">stderr</th>
<th align="right">Zscore</th>
<th align="right">BABA</th>
<th align="right">ABBA</th>
<th align="right">nsnps</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">French</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.0313</td>
<td align="right">0.006933</td>
<td align="right">4.510</td>
<td align="right">15802</td>
<td align="right">14844</td>
<td align="right">487753</td>
</tr>
<tr class="even">
<td align="left">Sardinian</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.0287</td>
<td align="right">0.006792</td>
<td align="right">4.222</td>
<td align="right">15729</td>
<td align="right">14852</td>
<td align="right">487646</td>
</tr>
<tr class="odd">
<td align="left">Han</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.0278</td>
<td align="right">0.006609</td>
<td align="right">4.199</td>
<td align="right">15780</td>
<td align="right">14928</td>
<td align="right">487925</td>
</tr>
<tr class="even">
<td align="left">Papuan</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.0457</td>
<td align="right">0.006571</td>
<td align="right">6.953</td>
<td align="right">16131</td>
<td align="right">14721</td>
<td align="right">487694</td>
</tr>
<tr class="odd">
<td align="left">Khomani_San</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.0066</td>
<td align="right">0.006292</td>
<td align="right">1.051</td>
<td align="right">16168</td>
<td align="right">15955</td>
<td align="right">487564</td>
</tr>
<tr class="even">
<td align="left">Mbuti</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">-0.0005</td>
<td align="right">0.006345</td>
<td align="right">-0.074</td>
<td align="right">15751</td>
<td align="right">15766</td>
<td align="right">487642</td>
</tr>
</tbody>
</table>
<p>We can see that in addition to the input information, this <code>data.frame</code> contains additional columns:</p>
<ul>
<li>
<code>D</code> - <span class="math inline">\(D\)</span> statistic value</li>
<li>
<code>stderr</code> - standard error of the <span class="math inline">\(D\)</span> statistic calculated using the block jackknife</li>
<li>
<code>Zscore</code> - <span class="math inline">\(Z\)</span>-zscore value (number of standard errors the <span class="math inline">\(D\)</span> is from 0, i.e. how strongly do we reject the null hypothesis of no admixture)</li>
<li>
<code>BABA</code>/<code>ABBA</code> - counts of observed site patterns</li>
<li>
<code>nsnps</code> - number of SNPs used for a give calculation</li>
</ul>
<p>(Output tables from other <code>admixr</code> functions follow a very similar format.)</p>
<p>While we could certainly make some inferences by looking at the <span class="math inline">\(Z\)</span>-scores, tables in general are not the best representation of this kind of data, especially as the number of samples increases. This is how we can use the <a href="https://ggplot2.tidyverse.org"><code>ggplot2</code></a> package to plot the results:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="kw">ggplot</span>(result, <span class="kw">aes</span>(<span class="kw">fct_reorder</span>(W, D), D, <span class="dt">color =</span> <span class="kw">abs</span>(Zscore) <span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb12-3" data-line-number="3"><span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="dv">0</span>, <span class="dt">linetype =</span> <span class="dv">2</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb12-4" data-line-number="4"><span class="st">  </span><span class="kw">geom_errorbar</span>(<span class="kw">aes</span>(<span class="dt">ymin =</span> D <span class="op">-</span><span class="st"> </span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span>stderr, <span class="dt">ymax =</span> D <span class="op">+</span><span class="st"> </span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span>stderr))</a></code></pre></div>
<p><img src="tutorial_files/figure-html/d_plot-1.png" width="672"></p>
<p>(If you want to more know about data analysis using R, including plotting with ggplot2, I highly recommend <a href="http://r4ds.had.co.nz">this</a> free book.)</p>
<p>We can see that all three Africans have <span class="math inline">\(D\)</span> values not significantly different from 0, meaning that the data is consistent with the null hypothesis of no Neanderthal ancestry in Africans. On the other hand, the test rejects the null hypothesis for all non-Africans today, suggesting that Neanderthals admixed with the ancestors of present-day non-Africans. In fact, this is a similar test to the one that was used as evidence supporting the Neanderthal admixture hypothesis in the first place!</p>
</div>
<div id="f_4-statistic" class="section level2">
<h2 class="hasAnchor">
<a href="#f_4-statistic" class="anchor"></a><span class="math inline">\(f_4\)</span> statistic</h2>
<p>An alternative way of addressing the previous question is to use the <span class="math inline">\(f_4\)</span> statistic, which is very similar to <span class="math inline">\(D\)</span> statistic and can be calculated as:</p>
<p><span class="math display">\[ f_4 = \frac{\textrm{# BABA sites - # ABBA sites}}{\textrm{# sites}}\]</span></p>
<p>Again, significant departure of <span class="math inline">\(f_4\)</span> from 0 is informative about gene flow, in a way analogous to <span class="math inline">\(D\)</span> statistic.</p>
<p>To repeat the previous analysis using <span class="math inline">\(f_4\)</span> statistic, we can run:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/f4ratio.html">f4</a></span>(<span class="dt">W =</span> pops, <span class="dt">X =</span> <span class="st">"Yoruba"</span>, <span class="dt">Y =</span> <span class="st">"Vindija"</span>, <span class="dt">Z =</span> <span class="st">"Chimp"</span>, <span class="dt">prefix =</span> eigenstrat_prefix)</a></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw">head</span>(result)</a></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">W</th>
<th align="left">X</th>
<th align="left">Y</th>
<th align="left">Z</th>
<th align="right">f4</th>
<th align="right">stderr</th>
<th align="right">Zscore</th>
<th align="right">BABA</th>
<th align="right">ABBA</th>
<th align="right">nsnps</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">French</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.001965</td>
<td align="right">0.000437</td>
<td align="right">4.501</td>
<td align="right">15802</td>
<td align="right">14844</td>
<td align="right">487753</td>
</tr>
<tr class="even">
<td align="left">Sardinian</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.001798</td>
<td align="right">0.000427</td>
<td align="right">4.209</td>
<td align="right">15729</td>
<td align="right">14852</td>
<td align="right">487646</td>
</tr>
<tr class="odd">
<td align="left">Han</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.001746</td>
<td align="right">0.000418</td>
<td align="right">4.178</td>
<td align="right">15780</td>
<td align="right">14928</td>
<td align="right">487925</td>
</tr>
<tr class="even">
<td align="left">Papuan</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.002890</td>
<td align="right">0.000417</td>
<td align="right">6.924</td>
<td align="right">16131</td>
<td align="right">14721</td>
<td align="right">487694</td>
</tr>
<tr class="odd">
<td align="left">Khomani_San</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">0.000436</td>
<td align="right">0.000415</td>
<td align="right">1.051</td>
<td align="right">16168</td>
<td align="right">15955</td>
<td align="right">487564</td>
</tr>
<tr class="even">
<td align="left">Mbuti</td>
<td align="left">Yoruba</td>
<td align="left">Vindija</td>
<td align="left">Chimp</td>
<td align="right">-0.000030</td>
<td align="right">0.000410</td>
<td align="right">-0.074</td>
<td align="right">15751</td>
<td align="right">15766</td>
<td align="right">487642</td>
</tr>
</tbody>
</table>
<p>We can see by comparing this to the <span class="math inline">\(D\)</span> statistic result above that we can make the same conclusions.</p>
<p>You might be wondering why we have both <span class="math inline">\(f_4\)</span> and <span class="math inline">\(D\)</span> if they are so similar. The truth is that <span class="math inline">\(f_4\)</span> is, among other things, directly informative about the amount of shared genetic drift (“branch length”) between pairs of populations, which is, in many cases, a very useful theoretical property. Other than that, it’s often a matter of personal preference and so <code>admixr</code> provides separate functions for calculating both.</p>
</div>
<div id="f_4-ratio-statistic" class="section level2">
<h2 class="hasAnchor">
<a href="#f_4-ratio-statistic" class="anchor"></a><span class="math inline">\(f_4\)</span>-ratio statistic</h2>
<p>Now we know that non-Africans today carry <em>some</em> Neanderthal ancestry. But what if we want to know <em>how much</em> Neanderthal ancestry they have? What proportion of their genomes is of Neanderthal origin?</p>
<p>Unfortunately, we don’t have enough space here to explain all the details about the inner workings of <span class="math inline">\(f_4\)</span>-ratio statistic. However, in general, when we are interested in estimating the <em>proportion</em> of ancestry in a population <span class="math inline">\(X\)</span> coming some parental lineage <span class="math inline">\(B\)</span>, we can use a ratio of two <span class="math inline">\(f_4\)</span> statistics.</p>
<p><span class="math display">\[f_4\textrm{-ratio} = \frac{f_4(A, O; X, C}{f_4(A, O; B, C}.\]</span></p>
<p>Using <code>amidxr</code>, we can calculate <span class="math inline">\(f_4\)</span>-ratios using the following code (<code>X</code> being a vector of samples for which we want to estimate Neanderthal ancestry):</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1">result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/f4ratio.html">f4ratio</a></span>(<span class="dt">X =</span> pops, <span class="dt">A =</span> <span class="st">"Altai"</span>, <span class="dt">B =</span> <span class="st">"Vindija"</span>, <span class="dt">C =</span> <span class="st">"Yoruba"</span>, <span class="dt">O =</span> <span class="st">"Chimp"</span>, <span class="dt">prefix =</span> eigenstrat_prefix)</a></code></pre></div>
<p>The ancestry proportion (a number between 0 and 1) is given in the <code>alpha</code> column:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="kw">head</span>(result)</a></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">A</th>
<th align="left">B</th>
<th align="left">X</th>
<th align="left">C</th>
<th align="left">O</th>
<th align="right">alpha</th>
<th align="right">stderr</th>
<th align="right">Zscore</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">French</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.023774</td>
<td align="right">0.006173</td>
<td align="right">3.851</td>
</tr>
<tr class="even">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Sardinian</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.024468</td>
<td align="right">0.006079</td>
<td align="right">4.025</td>
</tr>
<tr class="odd">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Han</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.022117</td>
<td align="right">0.005901</td>
<td align="right">3.748</td>
</tr>
<tr class="even">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Papuan</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.037311</td>
<td align="right">0.005821</td>
<td align="right">6.410</td>
</tr>
<tr class="odd">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Khomani_San</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.003909</td>
<td align="right">0.005923</td>
<td align="right">0.660</td>
</tr>
<tr class="even">
<td align="left">Altai</td>
<td align="left">Vindija</td>
<td align="left">Mbuti</td>
<td align="left">Yoruba</td>
<td align="left">Chimp</td>
<td align="right">0.000319</td>
<td align="right">0.005721</td>
<td align="right">0.056</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="kw">ggplot</span>(result, <span class="kw">aes</span>(<span class="kw">fct_reorder</span>(X, alpha), alpha, <span class="dt">color =</span> <span class="kw">abs</span>(Zscore) <span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb17-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb17-3" data-line-number="3"><span class="st">  </span><span class="kw">geom_errorbar</span>(<span class="kw">aes</span>(<span class="dt">ymin =</span> alpha <span class="op">-</span><span class="st"> </span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span>stderr, <span class="dt">ymax =</span> alpha <span class="op">+</span><span class="st"> </span><span class="dv">2</span> <span class="op">*</span><span class="st"> </span>stderr)) <span class="op">+</span></a>
<a class="sourceLine" id="cb17-4" data-line-number="4"><span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="dv">0</span>, <span class="dt">linetype =</span> <span class="dv">2</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb17-5" data-line-number="5"><span class="st">  </span><span class="kw">labs</span>(<span class="dt">y =</span> <span class="st">"Neandertal ancestry proportion"</span>, <span class="dt">x =</span> <span class="st">"present-day individual"</span>)</a></code></pre></div>
<p><img src="tutorial_files/figure-html/f4ratio_plot-1.png" width="672"></p>
<p>We can make several observations:</p>
<ul>
<li>Again, we don’t see any significant Neanderthal ancestry in present-day Africans (proportion is consistent with 0%), which is what we confirmed using <span class="math inline">\(D\)</span> and <span class="math inline">\(f_4\)</span> above.</li>
<li>Present-day non-Africans carry between 2-3% of Neanderthal ancestry.</li>
<li>We see a much higher proportion of Neanderthal ancestry in people from Papua New Guinea - more than 4%! This is consistent with earlier studies that suggest additional archaic admixture events in the ancestors of present-day Papuans.</li>
</ul>
</div>
<div id="f_3-statistic" class="section level2">
<h2 class="hasAnchor">
<a href="#f_3-statistic" class="anchor"></a><span class="math inline">\(f_3\)</span> statistic</h2>
<p>The <span class="math inline">\(f_3\)</span> statistic, also known as the 3-population statistic, is useful whenever we want to:</p>
<ol style="list-style-type: decimal">
<li>Estimate the branch length (shared genetic drift) between a pair of populations <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span> with respect to a common outgroup <span class="math inline">\(C\)</span>. In this case, the higher the <span class="math inline">\(f_3\)</span> value, the longer the shared evolutionary time between <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span>.</li>
<li>Test whether population <span class="math inline">\(C\)</span> is a mixture of two parental populations <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span>. Negative value of the <span class="math inline">\(f_3\)</span> statistic then serves as statistical evidence of this admixture.</li>
</ol>
<p>As an example, imagine we are interested in relative divergence times between pairs of present-day human populations, and want to know in which approximate order they split of from each other. To address this problem, we could use <span class="math inline">\(f_3\)</span> statistic by fixing the <span class="math inline">\(C\)</span> outgroup as Chimp, and calculating pairwise <span class="math inline">\(f_3\)</span> statistics between all pairs of present-day modern humans.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1">pops &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">"French"</span>, <span class="st">"Sardinian"</span>, <span class="st">"Han"</span>, <span class="st">"Papuan"</span>, <span class="st">"Khomani_San"</span>, <span class="st">"Mbuti"</span>, <span class="st">"Dinka"</span>, <span class="st">"Yoruba"</span>)</a>
<a class="sourceLine" id="cb18-2" data-line-number="2"></a>
<a class="sourceLine" id="cb18-3" data-line-number="3">result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/f4ratio.html">f3</a></span>(<span class="dt">A =</span> pops, <span class="dt">B =</span> pops, <span class="dt">C =</span> <span class="st">"Chimp"</span>, <span class="dt">prefix =</span> eigenstrat_prefix)</a></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="kw">head</span>(result)</a></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">A</th>
<th align="left">B</th>
<th align="left">C</th>
<th align="right">f3</th>
<th align="right">stderr</th>
<th align="right">Zscore</th>
<th align="right">nsnps</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">French</td>
<td align="left">Sardinian</td>
<td align="left">Chimp</td>
<td align="right">1.738202e+15</td>
<td align="right">4.373915e+12</td>
<td align="right">397.402</td>
<td align="right">225492</td>
</tr>
<tr class="even">
<td align="left">French</td>
<td align="left">Han</td>
<td align="left">Chimp</td>
<td align="right">1.658993e+15</td>
<td align="right">4.153303e+12</td>
<td align="right">399.440</td>
<td align="right">229514</td>
</tr>
<tr class="odd">
<td align="left">French</td>
<td align="left">Papuan</td>
<td align="left">Chimp</td>
<td align="right">1.626844e+15</td>
<td align="right">4.323632e+12</td>
<td align="right">376.268</td>
<td align="right">226981</td>
</tr>
<tr class="even">
<td align="left">French</td>
<td align="left">Khomani_San</td>
<td align="left">Chimp</td>
<td align="right">1.221558e+15</td>
<td align="right">3.667817e+12</td>
<td align="right">333.048</td>
<td align="right">243884</td>
</tr>
<tr class="odd">
<td align="left">French</td>
<td align="left">Mbuti</td>
<td align="left">Chimp</td>
<td align="right">1.259897e+15</td>
<td align="right">3.474916e+12</td>
<td align="right">362.569</td>
<td align="right">253162</td>
</tr>
<tr class="even">
<td align="left">French</td>
<td align="left">Dinka</td>
<td align="left">Chimp</td>
<td align="right">1.410976e+15</td>
<td align="right">3.752041e+12</td>
<td align="right">376.055</td>
<td align="right">257987</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="co"># sort the population labels according to an increasing f3 value relative to French</span></a>
<a class="sourceLine" id="cb20-2" data-line-number="2">ordered &lt;-<span class="st"> </span><span class="kw">filter</span>(result, A <span class="op">==</span><span class="st"> "French"</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">arrange</span>(f3) <span class="op">%&gt;%</span><span class="st"> </span>.[[<span class="st">"B"</span>]] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">"French"</span>)</a>
<a class="sourceLine" id="cb20-3" data-line-number="3"></a>
<a class="sourceLine" id="cb20-4" data-line-number="4"><span class="co"># plot heatmap of pairwise f3 values</span></a>
<a class="sourceLine" id="cb20-5" data-line-number="5">result <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb20-6" data-line-number="6"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">A =</span> <span class="kw">factor</span>(A, <span class="dt">levels =</span> ordered),</a>
<a class="sourceLine" id="cb20-7" data-line-number="7">         <span class="dt">B =</span> <span class="kw">factor</span>(B, <span class="dt">levels =</span> ordered)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb20-8" data-line-number="8"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(A, B)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_tile</span>(<span class="kw">aes</span>(<span class="dt">fill =</span> f3))</a></code></pre></div>
<p><img src="tutorial_files/figure-html/f3_plot-1.png" width="768"></p>
<p>We can see that when we order the heatmap labels based on values of pairwise <span class="math inline">\(f_3\)</span> statistics, the (already known) order of population splits pops up beautifully (i.e. San separated first, followed by Mbuti, etc.).</p>
</div>
<div id="qpadm-method" class="section level2">
<h2 class="hasAnchor">
<a href="#qpadm-method" class="anchor"></a>qpAdm method</h2>
<p>The last ADMIXTOOLS method implemented in <code>admixr</code> is qpAdm. Unfortunately, it is also one that is the most complex and has not been properly described and peer-reviewed yet. Nevertheless, it seems to have a lot of power to disentangle complex admixture scenarios, and so we included it in our package as well.</p>
<p>Very briefly, qpAdm can be used to estimate admixture proportions coming from a series of <span class="math inline">\(N\)</span> source <em>ancestral</em> populations, assuming we have reference populations that form clades with those <em>source</em> populations that are closer to them than to any of the specified <em>outgroup</em> populations.</p>
<p>Formally, if a Test population has ancestry coming from <span class="math inline">\(N\)</span> ancestral source populations, with Reference populations being closer to them than are outgroup populations <span class="math inline">\(O_i\)</span>, we can write:</p>
<p><span class="math display">\[f_4(\textrm{Test}, O_a, O_b, O_c) \approx \sum_{i=1}^N \alpha_i f_4(\textrm{Reference}_i, O_a; O_b, O_c),\]</span></p>
<p>where <span class="math inline">\(\sum_{i=1}^N \alpha_i = 1\)</span> and <span class="math inline">\(\alpha_i \geq 1\)</span> for all <span class="math inline">\(i = 1, ..., N\)</span>.</p>
<p>If this looks like black magic to you, I feel your pain and direct you to the Supplementary Section 9 of <a href="http://www.nature.com/articles/nature14317">Haak et al. 2015</a>, and an <a href="https://github.com/DReichLab/AdmixTools/blob/master/pdoc.pdf">informal write-up</a> distributed with the ADMIXTOOLS software. While the method does not have a dedicated paper yet, it’s not so complicated to understand the basic principles.</p>
<p>Probably the simplest possible case to show that qpAdm works is by returning to the question of estimating Neanderthal ancestry proportions. Let’s define:</p>
<ul>
<li>Europeans as the <em>target</em> samples to estimate ancestry proportions for</li>
<li>Vindija Neanderthal and an African as two <em>reference</em> populations (two potential sources of ancestries in Europeans today)</li>
<li>
<em>outgroup</em> populations - Chimp, Altai Neanderthal and Denisovan (which are all further from the true ancestral populations - Vindija and African - than the <em>reference</em> populations)</li>
</ul>
<p>Assuming all of that, we can run qpAdm with:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1">result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/qpAdm.html">qpAdm</a></span>(</a>
<a class="sourceLine" id="cb21-2" data-line-number="2">  <span class="dt">target =</span> <span class="kw">c</span>(<span class="st">"French"</span>, <span class="st">"Sardinian"</span>, <span class="st">"Mbuti"</span>, <span class="st">"Dinka"</span>),</a>
<a class="sourceLine" id="cb21-3" data-line-number="3">  <span class="dt">references =</span> <span class="kw">c</span>(<span class="st">"Vindija"</span>, <span class="st">"Yoruba"</span>),</a>
<a class="sourceLine" id="cb21-4" data-line-number="4">  <span class="dt">outgroups =</span> <span class="kw">c</span>(<span class="st">"Chimp"</span>, <span class="st">"Denisova"</span>, <span class="st">"Altai"</span>),</a>
<a class="sourceLine" id="cb21-5" data-line-number="5">  <span class="dt">prefix =</span> eigenstrat_prefix</a>
<a class="sourceLine" id="cb21-6" data-line-number="6">)</a></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1">result</a></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">target</th>
<th align="right">nsnps</th>
<th align="right">Vindija</th>
<th align="right">Yoruba</th>
<th align="right">stderr_Vindija</th>
<th align="right">stderr_Yoruba</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">French</td>
<td align="right">499434</td>
<td align="right">0.0215749</td>
<td align="right">0.9784251</td>
<td align="right">0.006</td>
<td align="right">0.006</td>
</tr>
<tr class="even">
<td align="left">Sardinian</td>
<td align="right">499314</td>
<td align="right">0.0246924</td>
<td align="right">0.9753076</td>
<td align="right">0.006</td>
<td align="right">0.006</td>
</tr>
<tr class="odd">
<td align="left">Mbuti</td>
<td align="right">499334</td>
<td align="right">0.0009431</td>
<td align="right">0.9990569</td>
<td align="right">0.006</td>
<td align="right">0.006</td>
</tr>
<tr class="even">
<td align="left">Dinka</td>
<td align="right">499362</td>
<td align="right">-0.0027000</td>
<td align="right">1.0027000</td>
<td align="right">0.005</td>
<td align="right">0.005</td>
</tr>
</tbody>
</table>
<p>If we compare this result to the <span class="math inline">\(f_4\)</span>-ratio values calculated above, we see that the qpAdm estimates are very close to what we got earlier.</p>
</div>
<div id="merging-population-labels" class="section level2">
<h2 class="hasAnchor">
<a href="#merging-population-labels" class="anchor"></a>Merging population labels</h2>
<p>What we’ve been doing so far was calculating statistics for individual samples. However, it is often useful to treat multiple samples as a single group or population. <code>admixr</code> provides a function called <code>group_labels</code> that does just that.</p>
<p>Here is an example: let’s say we want to run a similar analysis to the one described in the <span class="math inline">\(D\)</span> statistic section, but we want to treat Europeans, Africans and archaics combined populations. But the <code>ind</code> file that we have does not contain grouped labels - each sample stands on its own:</p>
<pre><code>Chimp        U  Chimp
Mbuti        U  Mbuti
Yoruba       U  Yoruba
Khomani_San  U  Khomani_San
Han          U  Han
Dinka        U  Dinka
Sardinian    U  Sardinian
Papuan       U  Papuan
French       U  French
Vindija      U  Vindija
Altai        U  Altai
Denisova     U  Denisova</code></pre>
<p>To merge several individual samples under a combined label we can call <code>group_labels</code> like this:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1"><span class="co"># paths to the original ind file and a new modified ind file, which will</span></a>
<a class="sourceLine" id="cb24-2" data-line-number="2"><span class="co"># contain merged population labels</span></a>
<a class="sourceLine" id="cb24-3" data-line-number="3">ind_path &lt;-<span class="st"> </span><span class="kw">paste0</span>(eigenstrat_prefix, <span class="st">".ind"</span>)</a>
<a class="sourceLine" id="cb24-4" data-line-number="4">modif_path &lt;-<span class="st"> </span><span class="kw">tempfile</span>()</a>
<a class="sourceLine" id="cb24-5" data-line-number="5"></a>
<a class="sourceLine" id="cb24-6" data-line-number="6"><span class="kw"><a href="../reference/group_labels.html">group_labels</a></span>(</a>
<a class="sourceLine" id="cb24-7" data-line-number="7">  <span class="dt">ind =</span> ind_path,</a>
<a class="sourceLine" id="cb24-8" data-line-number="8">  <span class="dt">modified_ind =</span> modif_path,</a>
<a class="sourceLine" id="cb24-9" data-line-number="9">  <span class="dt">labels =</span> <span class="kw">list</span>( <span class="co"># new population labels</span></a>
<a class="sourceLine" id="cb24-10" data-line-number="10">    <span class="dt">European =</span> <span class="kw">c</span>(<span class="st">"French"</span>, <span class="st">"Sardinian"</span>),</a>
<a class="sourceLine" id="cb24-11" data-line-number="11">    <span class="dt">African =</span> <span class="kw">c</span>(<span class="st">"Dinka"</span>, <span class="st">"Yoruba"</span>, <span class="st">"Mbuti"</span>, <span class="st">"Khomani_San"</span>),</a>
<a class="sourceLine" id="cb24-12" data-line-number="12">    <span class="dt">Archaic =</span> <span class="kw">c</span>(<span class="st">"Vindija"</span>, <span class="st">"Altai"</span>, <span class="st">"Denisova"</span>)</a>
<a class="sourceLine" id="cb24-13" data-line-number="13">  )</a>
<a class="sourceLine" id="cb24-14" data-line-number="14">)</a></code></pre></div>
<p>This is what a modified <code>ind</code> file generated by <code>group_labels</code> looks like:</p>
<pre><code>Chimp        U  Chimp
Mbuti        U  African
Yoruba       U  African
Khomani_San  U  African
Han          U  Han
Dinka        U  African
Sardinian    U  European
Papuan       U  Papuan
French       U  European
Vindija      U  Archaic
Altai        U  Archaic
Denisova     U  Archaic</code></pre>
<p>We can then use “European”, “African” and “Archaic” labels in any of the <code>admixr</code> wrapper functions described above, we just have to specify a new <code>ind</code> file in addition to the <code>prefix</code> argument. For example:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb26-1" data-line-number="1">result &lt;-<span class="st"> </span><span class="kw"><a href="../reference/f4ratio.html">d</a></span>(<span class="dt">W =</span> <span class="st">"European"</span>, <span class="dt">X =</span> <span class="st">"African"</span>, <span class="dt">Y =</span> <span class="st">"Archaic"</span>, <span class="dt">Z =</span> <span class="st">"Chimp"</span>,</a>
<a class="sourceLine" id="cb26-2" data-line-number="2">            <span class="dt">prefix =</span> eigenstrat_prefix, <span class="dt">ind =</span> modif_path)</a></code></pre></div>
<p>Here is the result, showing (as we’ve seen above for individual samples) that Europeans show genetic affinity to archaic humans compared to Africans today:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1"><span class="kw">head</span>(result)</a></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">W</th>
<th align="left">X</th>
<th align="left">Y</th>
<th align="left">Z</th>
<th align="right">D</th>
<th align="right">stderr</th>
<th align="right">Zscore</th>
<th align="right">BABA</th>
<th align="right">ABBA</th>
<th align="right">nsnps</th>
</tr></thead>
<tbody><tr class="odd">
<td align="left">European</td>
<td align="left">African</td>
<td align="left">Archaic</td>
<td align="left">Chimp</td>
<td align="right">0.0225</td>
<td align="right">0.004404</td>
<td align="right">5.117</td>
<td align="right">15487</td>
<td align="right">14805</td>
<td align="right">489003</td>
</tr></tbody>
</table>
<p>Note that in the <code><a href="../reference/f4ratio.html">d()</a></code> call we provided both path to shared EIGENSTRAT prefix, but we also specified the path to the modified <code>ind</code> file separately. This overides the unmodified <code>ind</code> file that would be normally used, and ADMIXTOOLS picks up the combined population labels from the modified <code>ind</code> file.</p>
</div>
<div id="merging-eigenstrat-datasets" class="section level2">
<h2 class="hasAnchor">
<a href="#merging-eigenstrat-datasets" class="anchor"></a>Merging EIGENSTRAT datasets</h2>
<p>Another useful data processing function is <code>merge_eigenstrat</code>. This function takes two EIGENSTRAT datasets (i.e. trios of ind/snp/geno files) and merges them, producing a union of samples and intersection of SNPs from both of them.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb28-1" data-line-number="1"><span class="co"># this is just an example code - it will not run unless you specify the paths</span></a>
<a class="sourceLine" id="cb28-2" data-line-number="2"><span class="kw"><a href="../reference/merge_eigenstrat.html">merge_eigenstrat</a></span>(</a>
<a class="sourceLine" id="cb28-3" data-line-number="3">    <span class="dt">prefix =</span> output_prefix,       <span class="co"># prefix of the final combined dataset</span></a>
<a class="sourceLine" id="cb28-4" data-line-number="4">    <span class="dt">input1 =</span> path_to_1st_dataset, <span class="co"># prefix of the first dataset</span></a>
<a class="sourceLine" id="cb28-5" data-line-number="5">    <span class="dt">input2 =</span> path_to_2nd_dataset  <span class="co"># prefix of the first dataset</span></a>
<a class="sourceLine" id="cb28-6" data-line-number="6">)</a></code></pre></div>
</div>
<div id="counting-presentmissing-snps" class="section level2">
<h2 class="hasAnchor">
<a href="#counting-presentmissing-snps" class="anchor"></a>Counting present/missing SNPs</h2>
<p>The <code>count_snps</code> function can be useful for quality control, weighting of admixture statistics (<span class="math inline">\(D\)</span>, <span class="math inline">\(f_4\)</span>, etc.) for regression analyses etc. There are two optional arguments:</p>
<ul>
<li>
<code>prop</code> - changes whether to report SNP counts or proportions (set to <code>FALSE</code> by default),</li>
<li>
<code>missing</code> - controls whether to count missing SNPs instead of present SNPs (set to <code>FALSE</code> by default).</li>
</ul>
<p>For each sample, count the SNPs present in that sample:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1"><span class="kw"><a href="../reference/count_snps.html">count_snps</a></span>(eigenstrat_prefix)</a></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">name</th>
<th align="right">present</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Chimp</td>
<td align="right">491273</td>
</tr>
<tr class="even">
<td align="left">Mbuti</td>
<td align="right">499334</td>
</tr>
<tr class="odd">
<td align="left">Yoruba</td>
<td align="right">499246</td>
</tr>
<tr class="even">
<td align="left">Khomani_San</td>
<td align="right">499250</td>
</tr>
<tr class="odd">
<td align="left">Han</td>
<td align="right">499654</td>
</tr>
<tr class="even">
<td align="left">Dinka</td>
<td align="right">499362</td>
</tr>
<tr class="odd">
<td align="left">Sardinian</td>
<td align="right">499314</td>
</tr>
<tr class="even">
<td align="left">Papuan</td>
<td align="right">499377</td>
</tr>
<tr class="odd">
<td align="left">French</td>
<td align="right">499434</td>
</tr>
<tr class="even">
<td align="left">Vindija</td>
<td align="right">497544</td>
</tr>
<tr class="odd">
<td align="left">Altai</td>
<td align="right">497729</td>
</tr>
<tr class="even">
<td align="left">Denisova</td>
<td align="right">497398</td>
</tr>
</tbody>
</table>
</div>
<div id="data-filtering" class="section level2">
<h2 class="hasAnchor">
<a href="#data-filtering" class="anchor"></a>Data filtering</h2>
<div id="filtering-based-on-a-bed-file" class="section level3">
<h3 class="hasAnchor">
<a href="#filtering-based-on-a-bed-file" class="anchor"></a>Filtering based on a BED file</h3>
<p>A common situation in genomics is performing an analysis on a subset of the genome. However, EIGENSTRAT is a rather obscure file format which makes it very difficult to find bioinformatics tools that support it. Luckily, <code>admixr</code> includes a function <code><a href="../reference/filter_sites.html">filter_sites()</a></code> that takes an EIGENSTRAT prefix and a BED file and produces a snp file, that contains simply contains positions of sites that either overlap (or don’t overlap) regions/sites specified in the BED file.</p>
<p>Why is that useful? Each of the main <code>admixr</code> functions accepts an <code>exclude =</code> argument, which specifies a path to a snp file with positions of we want to exclude from our analysis. This is file is exactly what <code><a href="../reference/filter_sites.html">filter_sites()</a></code> function generates.</p>
<p>This might seem like a very convoluted way to do this type of analysis but this is how ADMIXTOOLS actually works under the hood, and being “just” a wrapper around ADMIXTOOLS we have no other choice but to comply. There is a rather obscure option called <a href="https://github.com/DReichLab/AdmixTools/blob/master/README.ROLLOFF#L48"><code>badsnpfile</code></a>, which specifies the sites to exclude from the calculation. The <code>exclude =</code> argument of <code>admixr</code> functions simply fills in the <code>badsnpfile</code> parameter before starting an ADMIXTOOLS command.</p>
<p>(An alternative to this way of filtering data would be to create a whole new EIGENSTRAT snp/geno/ind trio for each new subset. However, this is problematic if the data set is huge, or if we need to repeat an analysis on many different subsets of data. In my experience, I found that generating just a snp file instead of creating a new subset of the complete data is a good compromise. If you really think this package should implement this alternative, please get in touch and we can talk about it.)</p>
<p>Here are two examples how to filter EIGENSTRAT data (just for illustration, they will not run by themselves without specifying all the paths):</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb30-1" data-line-number="1"><span class="co"># run this if your BED file contains regions to keep in an analysis</span></a>
<a class="sourceLine" id="cb30-2" data-line-number="2"><span class="kw"><a href="../reference/filter_sites.html">filter_sites</a></span>(<span class="dt">prefix =</span> eigenstrat_prefix, <span class="dt">bed =</span> regions_to_include, <span class="dt">outsnp =</span> excluded_snps1)</a>
<a class="sourceLine" id="cb30-3" data-line-number="3"></a>
<a class="sourceLine" id="cb30-4" data-line-number="4"><span class="co"># run this if your BED file contains regions to remove from an analysis</span></a>
<a class="sourceLine" id="cb30-5" data-line-number="5"><span class="kw"><a href="../reference/filter_sites.html">filter_sites</a></span>(<span class="dt">prefix =</span> eigenstrat_prefix, <span class="dt">bed =</span> regions_to_exclude, <span class="dt">outsnp =</span> excluded_snps2)</a></code></pre></div>
<p>Both commands do a different thing, but both generate a snp file that specifies which sites need to be excluded from an analysis. It can be used in this way:</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb31-1" data-line-number="1"><span class="kw"><a href="../reference/f4ratio.html">d</a></span>(pop_W, pop_X, pop_Y, pop_Z, <span class="dt">prefix =</span> eigenstrat_prefix, <span class="dt">exclude =</span> excluded_snps1)</a></code></pre></div>
</div>
<div id="filtering-out-potential-ancient-dna-damage-snps" class="section level3">
<h3 class="hasAnchor">
<a href="#filtering-out-potential-ancient-dna-damage-snps" class="anchor"></a>Filtering out potential ancient DNA damage SNPs</h3>
<p>In an ancient DNA world, we often need to repeat an analysis on a subset of data that is less likely to be influenced by ancient DNA damage, to verify that our results are not caused by artifacts in the data (due to biochemical properties of DNA degradation, ancient DNA damage will lead to an increase in C→T and G→A substitutions). Using a similar method described in a general BED filtering section above, we can use the <code><a href="../reference/filter_damage.html">filter_damage()</a></code> function to generate a snp file with positions that carry transitions (C→T and G→A sites), that can be removed using the <code>exclude =</code> argument of the main <code>admixr</code> functions.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb32-1" data-line-number="1"><span class="co"># generate snp file with positions of transitions</span></a>
<a class="sourceLine" id="cb32-2" data-line-number="2">transitions &lt;-<span class="st"> </span><span class="kw">tempfile</span>()</a>
<a class="sourceLine" id="cb32-3" data-line-number="3"><span class="kw"><a href="../reference/filter_damage.html">filter_damage</a></span>(<span class="dt">prefix =</span> eigenstrat_prefix, <span class="dt">outsnp =</span> transitions)</a>
<a class="sourceLine" id="cb32-4" data-line-number="4"></a>
<a class="sourceLine" id="cb32-5" data-line-number="5"><span class="co"># perform the calculation only on transversions</span></a>
<a class="sourceLine" id="cb32-6" data-line-number="6"><span class="kw"><a href="../reference/f4ratio.html">d</a></span>(<span class="dt">W =</span> <span class="st">"French"</span>, <span class="dt">X =</span> <span class="st">"Dinka"</span>, <span class="dt">Y =</span> <span class="st">"Altai"</span>, <span class="dt">Z =</span> <span class="st">"Chimp"</span>, <span class="dt">prefix =</span> eigenstrat_prefix, <span class="dt">exclude =</span> transitions)</a></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#installation">Installation</a></li>
      <li><a href="#a-note-about-eigenstrat-format">A note about EIGENSTRAT format</a></li>
      <li><a href="#philosophy-of-admixr">Philosophy of <code>admixr</code></a></li>
      <li><a href="#d-statistic"><span class="math inline">\(D\)</span> statistic</a></li>
      <li><a href="#f_4-statistic"><span class="math inline">\(f_4\)</span> statistic</a></li>
      <li><a href="#f_4-ratio-statistic"><span class="math inline">\(f_4\)</span>-ratio statistic</a></li>
      <li><a href="#f_3-statistic"><span class="math inline">\(f_3\)</span> statistic</a></li>
      <li><a href="#qpadm-method">qpAdm method</a></li>
      <li><a href="#merging-population-labels">Merging population labels</a></li>
      <li><a href="#merging-eigenstrat-datasets">Merging EIGENSTRAT datasets</a></li>
      <li><a href="#counting-presentmissing-snps">Counting present/missing SNPs</a></li>
      <li><a href="#data-filtering">Data filtering</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Martin Petr.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
